// tegami/dist/cli/index.js
import { n as generateFromCommits, t as changelogFilename } from "../generate-Cip5o2xy.js";
import { i as handlePluginError, s as isCI, t as CancelledError } from "../error-CAsrGb6e.js";
import { n as renderChangelog } from "../shared-BraxbVVG.js";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path, { join, relative } from "node:path";
import { x } from "tinyexec";
import { autocompleteMultiselect, confirm, intro, isCancel, multiline, note, outro, select, spinner } from "@clack/prompts";
import { parseArgs } from "node:util";
//#region src/utils/git-changes.ts
async function getChangedPackages(packages, cwd) {
	const files = await getChangedFilePaths(cwd);
	const sortedPackages = packages.toSorted((a, b) => b.path.length - a.path.length);
	const matched = /* @__PURE__ */ new Set();
	for (const file of files) {
		const fullPath = join(cwd, file);
		for (const pkg of sortedPackages) if (!relative(pkg.path, fullPath).startsWith("..")) {
			matched.add(pkg);
			break;
		}
	}
	return matched;
}
async function getChangedFilePaths(cwd) {
	const files = /* @__PURE__ */ new Set();
	await Promise.all([["diff", "--name-only"], [
		"diff",
		"--cached",
		"--name-only"
	]].map(async (args) => {
		const result = await x("git", args, { nodeOptions: { cwd } });
		if (result.exitCode !== 0) return;
		for (const line of result.stdout.split("\n")) {
			const trimmed = line.trim();
			if (trimmed) files.add(trimmed);
		}
	}));
	return Array.from(files);
}
//#endregion
//#region src/cli/changelog.ts
async function runChangelogTui(tegami) {
	const context = await tegami._internal.context();
	intro("Create changelogs");
	let selectedPackages = [];
	if (!isCI()) selectedPackages = await promptPackageSelection(context.graph, context.cwd);
	if (selectedPackages.length === 0) {
		if (!isCI()) {
			const confirmed = await confirm({
				message: "Auto-generate changelog files from commits?",
				initialValue: true
			});
			if (isCancel(confirmed)) throw new CancelledError();
			if (!confirmed) {
				outro("No changelogs created.");
				return;
			}
		}
		await persistChangelogs(context, (await generateFromCommits(context)).map(({ filename, content, packages }) => ({
			filename,
			content,
			packages
		})), "No matching conventional commits were found.");
		return;
	}
	const packageBumpMap = await promptPackageBumpTypes(selectedPackages);
	const message = await multiline({
		message: "Describe change (Markdown supported, press tab then enter to exit)",
		placeholder: "The first line is heading\n\nAdditional description.",
		showSubmit: true,
		validate(value) {
			if (!value?.trim()) return "Enter a message.";
		}
	});
	if (isCancel(message)) throw new CancelledError();
	await persistChangelogs(context, [{
		filename: changelogFilename(),
		content: renderChangelog({ packages: packageBumpMap }, `## ${message.trim()}`),
		packages: packageBumpMap
	}]);
}
async function persistChangelogs(context, entries, emptyMessage = "No changelogs created.") {
	const s = spinner();
	s.start("Creating changelog");
	await mkdir(context.changelogDir, { recursive: true });
	await Promise.all(entries.map(({ filename, content }) => writeFile(join(context.changelogDir, filename), content)));
	s.stop(entries.length === 1 ? "Created 1 changelog file" : entries.length > 0 ? `Created ${entries.length} changelog files` : "No changelogs created");
	if (entries.length === 0) note(emptyMessage, "No changelogs created");
	else {
		const lines = [];
		for (const { filename, packages } of entries) {
			lines.push(filename);
			for (const [name, config] of Object.entries(packages)) {
				if (typeof config === "string") {
					lines.push(`${name}: ${config}`);
					continue;
				}
				if (config.replay?.length) {
					lines.push(`${name}: ${config.type} (replay on ${config.replay.join(" or ")})`);
					continue;
				}
				lines.push(`${name}: ${config.type}`);
			}
		}
		note(lines.join("\n"), "Created changelogs");
	}
	outro(entries.length === 1 ? "Changelog ready." : "Changelogs ready.");
}
async function promptPackageSelection(graph, cwd) {
	const useShortname = /* @__PURE__ */ new Map();
	const versionablePackages = graph.getPackages().filter((pkg) => pkg.version !== void 0);
	for (const pkg of versionablePackages) if (useShortname.has(pkg.name)) useShortname.set(pkg.name, false);
	else useShortname.set(pkg.name, true);
	const getPackageLabel = (pkg) => {
		return useShortname.get(pkg.name) ? pkg.name : pkg.id;
	};
	const changedPackages = await getChangedPackages(versionablePackages, cwd);
	const selectOptions = [];
	const groups = [];
	for (const group of graph.getGroups()) {
		const changed = group.packages.some((pkg) => changedPackages.has(pkg));
		groups.push([group, changed]);
	}
	groups.sort((a, b) => (a[1] ? 0 : 1) - (b[1] ? 0 : 1));
	for (const [group, changed] of groups) selectOptions.push({
		label: `(Group) ${group.name}` + (changed ? "*" : ""),
		value: `group:${group.name}`,
		hint: group.packages.map(getPackageLabel).join(", ")
	});
	for (const pkg of versionablePackages.sort((a, b) => (changedPackages.has(a) ? 0 : 1) - (changedPackages.has(b) ? 0 : 1))) selectOptions.push({
		label: getPackageLabel(pkg) + (changedPackages.has(pkg) ? "*" : ""),
		value: pkg.id
	});
	const selected = await autocompleteMultiselect({
		message: "Select packages (leave empty to auto-generate from commits)",
		required: false,
		options: selectOptions,
		initialValues: Array.from(changedPackages, (pkg) => pkg.id)
	});
	if (isCancel(selected)) throw new CancelledError();
	return selected;
}
async function promptPackageBumpTypes(selectedPackages) {
	const packageBumpMap = {};
	const bumpType = await select({
		message: "Select release type",
		options: [
			{
				value: "patch",
				label: "patch"
			},
			{
				value: "minor",
				label: "minor"
			},
			{
				value: "major",
				label: "major"
			},
			{
				value: "per-package",
				label: "choose per-package"
			}
		]
	});
	if (isCancel(bumpType)) throw new CancelledError();
	if (bumpType === "per-package") for (const pkg of selectedPackages) {
		const selectedBump = await select({
			message: `Select release type for "${pkg}"`,
			options: [
				{
					value: "patch",
					label: "patch"
				},
				{
					value: "minor",
					label: "minor"
				},
				{
					value: "major",
					label: "major"
				}
			]
		});
		if (isCancel(selectedBump)) throw new CancelledError();
		packageBumpMap[pkg] = selectedBump;
	}
	else for (const pkg of selectedPackages) packageBumpMap[pkg] = bumpType;
	return packageBumpMap;
}
//#endregion
//#region src/cli/init-agent.ts
async function initAgent(context, options = {}) {
	const output = path.resolve(context.cwd, options.output ?? "AGENTS.md");
	const content = renderAgentsMd(context);
	try {
		await writeFile(output, (await readFile(output, "utf8")).trimEnd() + "\n\n" + content);
		return {
			path: output,
			created: false
		};
	} catch {
		await writeFile(output, content);
		return {
			path: output,
			created: true
		};
	}
}
function renderAgentsMd(context) {
	const changelogDir = path.relative(context.cwd, context.changelogDir) || "project root";
	const lockPath = path.relative(context.cwd, context.lockPath) || "project root";
	return [
		"# Release workflow",
		"",
		"This repository uses [Tegami](https://tegami.fuma-nama.dev) for versioning and publishing.",
		"",
		"## Write changelog files",
		"",
		`Create pending changelog files under \`${changelogDir}/\` as \`YYYY-MM-DD-{hash}.md\`.`,
		"",
		`See the [changelog format docs](https://tegami.fuma-nama.dev/changelog) for details.`,
		"",
		"### Example",
		"",
		"```md",
		"---",
		"packages:",
		`  "npm:@acme/ui": patch`,
		"---",
		"",
		"### Fix button hover state",
		"",
		"The hover color now matches the design system.",
		"```",
		"",
		"### Package references",
		"",
		"Use package names, ids, or groups in frontmatter. For example:",
		"",
		"- `\"@acme/ui\"` — package name",
		"- `\"npm:@acme/ui\"` — package id",
		"- `\"group:acme\"` — every package in a group",
		"",
		"Rules:",
		"",
		"- Include YAML frontmatter with `packages`",
		"- Include at least one `#`, `##`, or `###` heading in the body",
		"- Write user-facing release notes under each heading",
		`- Do not edit the publish lock file (\`${lockPath}\`) or package \`CHANGELOG.md\` files directly`,
		""
	].join("\n");
}
//#endregion
//#region src/cli/core.ts
function createTegamiCliRegistry(tegami) {
	const commands = /* @__PURE__ */ new Map();
	function getChildCommands(command) {
		const out = [];
		const prefix = command.path.join("\0");
		for (const [key, candidate] of commands) if (key.startsWith(prefix) && key.length > prefix.length) out.push(candidate);
		return out;
	}
	return {
		command(path, options) {
			const definition = {
				path: path.length === 0 ? [] : path.split(/\s+/),
				positionals: [],
				description: options?.description,
				options: [],
				resolve: options?.resolve ?? true
			};
			commands.set(definition.path.join("\0"), definition);
			const api = {
				option(name, { short, description = "", type, multiple }) {
					definition.options.push({
						name,
						short,
						type,
						multiple,
						description
					});
					return api;
				},
				positional(name, required) {
					definition.positionals.push({
						name,
						required: required ?? true
					});
					return api;
				},
				positionals(name) {
					definition.positionals.push({
						name,
						required: false,
						multiple: true
					});
					return api;
				},
				action(fn) {
					definition.action = fn;
				}
			};
			return api;
		},
		help: () => formatRootHelp(commands),
		async parse(argv) {
			if (argv[0] === "--help" || argv[0] === "-h") {
				console.log(formatRootHelp(commands));
				return;
			}
			const match = findCommand(commands, argv);
			if (!match) throw new Error(`Unknown command: ${argv[0]}`);
			const args = argv.slice(match.path.length);
			if (args.includes("--help") || args.includes("-h")) {
				console.log(formatCommandHelp(match, getChildCommands(match)));
				return;
			}
			if (!match.action) {
				if (args.length === 0) {
					console.log(formatCommandHelp(match, getChildCommands(match)));
					return;
				}
				throw new Error(`Unknown command: ${argv.slice(0, match.path.length + 1).join(" ")}`);
			}
			const { values, positionals } = parseCommandArgs(match, args);
			const context = match.resolve ? await tegami._internal.context() : await tegami._internal.contextUnresolved();
			await match.action({
				context,
				values,
				positionals
			});
		}
	};
}
function formatUsage(command) {
	const s = [...command.path];
	for (const positional of command.positionals) {
		let c = positional.multiple ? `...${positional.name}` : positional.name;
		s.push(positional.required ? `<${c}>` : `[${c}]`);
	}
	return s.join(" ");
}
function findCommand(commands, argv) {
	if (argv.length === 0) return commands.get("");
	for (let i = 0; i < argv.length; i++) {
		const search = argv.slice(0, argv.length - i).join("\0");
		const res = commands.get(search);
		if (res) return res;
	}
}
function parseCommandArgs(command, args) {
	const optionsConfig = { help: {
		type: "boolean",
		short: "h"
	} };
	for (const option of command.options) {
		const config = optionsConfig[option.name] = { type: option.type };
		if (option.short) config.short = option.short;
		if (option.multiple) config.multiple = true;
	}
	const parsed = parseArgs({
		args,
		options: optionsConfig,
		strict: true,
		allowPositionals: command.positionals.length > 0
	});
	const positionals = {};
	for (const positional of command.positionals) {
		if (positional.multiple) {
			const value = positionals[positional.name] = [];
			while (parsed.positionals.length > 0) value.push(parsed.positionals.shift());
			continue;
		}
		const value = parsed.positionals.shift();
		if (value === void 0 && positional.required) throw new Error(`missing required argument: ${positional.name}`);
		if (value === void 0) continue;
		positionals[positional.name] = value;
	}
	if (parsed.positionals.length > 0) throw new Error("Too many arguments");
	return {
		values: parsed.values,
		positionals
	};
}
function formatRootHelp(commands) {
	let width = 0;
	const rows = [];
	for (const command of commands.values()) {
		if (command.path.length === 0) continue;
		const key = formatUsage(command);
		width = Math.max(width, key.length);
		rows.push([key, command.description ?? ""]);
	}
	return `${[
		"Usage: tegami [command]",
		"",
		commands.get("")?.description ?? "create changelogs",
		"",
		"Commands:",
		...rows.map(([usage, description]) => `  ${usage.padEnd(width)}   ${description}`),
		"",
		"Run without a command to open the changelog TUI."
	].join("\n")}\n`;
}
function formatCommandHelp(command, childCommands) {
	let width = 0;
	let childWidth = 0;
	const optionRows = command.options.map((option) => {
		const short = option.short ? `-${option.short}, ` : "";
		const long = `--${option.name}`;
		const key = option.type === "string" ? `${short}${long} <value>` : `${short}${long}`;
		width = Math.max(width, key.length);
		return [key, option.description ?? ""];
	});
	const childRows = childCommands.map((child) => {
		const key = formatUsage(child);
		childWidth = Math.max(childWidth, key.length);
		return [key, child.description ?? ""];
	});
	const lines = [];
	if (command.action) lines.push(`Usage: ${formatUsage(command)}`, "");
	if (command.description) lines.push(command.description, "");
	if (optionRows.length > 0) {
		lines.push("Options:");
		for (const [flags, description] of optionRows) lines.push(`  ${flags.padEnd(width)}   ${description}`);
		lines.push("");
	}
	if (childRows.length > 0) {
		lines.push("Commands:");
		for (const [usage, description] of childRows) lines.push(`  ${usage.padEnd(childWidth)}   ${description}`);
		lines.push("");
	}
	return lines.join("\n");
}
//#endregion
//#region src/cli/index.ts
function createCli(tegami, options = {}) {
	const $cli = init();
	async function init() {
		const ctx = await tegami._internal.contextUnresolved();
		const cli = createTegamiCliRegistry(tegami);
		registerCoreCommands(cli, tegami, options);
		for (const plugin of ctx.plugins) await handlePluginError(plugin, "initCli", () => plugin.initCli?.call(ctx, cli));
		return cli;
	}
	return { async parseAsync(argv = process.argv.slice(2)) {
		try {
			await (await $cli).parse(argv);
		} catch (error) {
			if (error instanceof CancelledError) outro(error.message);
			else {
				note(error instanceof Error ? error.message : String(error), "Error");
				outro("Command failed.");
			}
			process.exit(1);
		}
	} };
}
async function runCli(tegami, options = {}) {
	await createCli(tegami, options).parseAsync();
}
function registerCoreCommands(cli, tegami, options) {
	cli.command("", { description: "create changelog files interactively" }).action(async () => {
		await runChangelogTui(tegami);
	});
	cli.command("version", { description: "draft version changes and write the publish lock" }).option("no-checks", {
		type: "boolean",
		description: "skip checking whether the publish lock is still pending"
	}).action(async ({ values }) => {
		await versionPackages(tegami, {
			cli: options,
			noChecks: values["no-checks"]
		});
	});
	cli.command("ci", { description: "version and publish packages" }).action(async () => {
		if (await versionPackages(tegami, {
			cli: options,
			noChecks: true
		})) return;
		await publishPackages(tegami, { cli: options });
	});
	cli.command("check-publish", { description: "exit with code 1 if no publishing needed, otherwise 0" }).action(async () => {
		const { status } = await tegami.getPublishStatus();
		process.exit(status === "pending" ? 0 : 1);
	});
	cli.command("publish", { description: "publish packages from the publish lock" }).option("dry-run", {
		type: "boolean",
		description: "validate the publish lock without publishing"
	}).positionals("packages").action(async ({ values, positionals }) => {
		await publishPackages(tegami, {
			dryRun: values["dry-run"],
			packages: positionals.packages?.length ? positionals.packages : void 0,
			cli: options
		});
	});
	cli.command("cleanup", { description: "remove the publish lock after all packages have been published" }).action(async () => {
		await runCleanup(tegami);
	});
	cli.command("init-agent", {
		description: "write AGENTS.md with changelog instructions for AI agents",
		resolve: false
	}).option("output", {
		type: "string",
		short: "o",
		description: "output path"
	}).action(async ({ values }) => {
		await runInitAgent(tegami, { output: values.output ?? "AGENTS.md" });
	});
}
async function versionPackages(tegami, options) {
	intro("Version Packages");
	const { version: customVersion } = options.cli;
	const context = await tegami._internal.context();
	const draft = customVersion ? await customVersion() : await tegami.draft();
	if (!draft.canApply()) throw new Error(`The draft from custom "version" hook must not be applied`);
	for (const plugin of context.plugins) await handlePluginError(plugin, "initCliDraft", () => plugin.initCliDraft?.call(context, draft));
	if (!draft.hasPending()) {
		note("No pending version changes matched workspace packages.", "Nothing to version");
		outro("No versions changed.");
		return false;
	}
	if (!options.noChecks) {
		const { status, reason } = await tegami.getPublishStatus();
		if (status === "pending") {
			note(`Publish lock at ${context.lockPath} is still pending. Publish it before applying a new draft.`, "Failed to apply");
			outro(reason ?? "Cannot apply.");
			return false;
		}
	}
	const lines = [];
	for (const pkg of context.graph.getPackages()) {
		const plan = draft.getPackageDraft(pkg.id);
		if (!plan) continue;
		const bumped = plan.bumpVersion(pkg);
		if (!pkg.version || !bumped || bumped === pkg.version) continue;
		lines.push(`${pkg.id}: ${pkg.version} → ${bumped} (${plan.changelogs?.length ?? 0} changelogs)`);
		if (plan.bumpReasons) for (const reason of plan.bumpReasons) lines.push(`  - ${reason}`);
	}
	note(lines.join("\n"), "Release plan");
	const s = spinner();
	s.start("Updating package versions");
	try {
		await draft.apply();
	} catch (error) {
		s.stop("Failed to apply draft");
		throw error;
	}
	s.stop("Package versions updated");
	for (const plugin of context.plugins) await handlePluginError(plugin, "applyCliDraft", () => plugin.applyCliDraft?.call(context, draft));
	outro("Publish lock written.");
	return true;
}
async function publishPackages(tegami, options) {
	const dryRun = options.dryRun ?? false;
	const context = await tegami._internal.context();
	const { publish: customPublish } = options.cli;
	intro(dryRun ? "Publish packages (dry run)" : "Publish packages");
	const s = spinner();
	s.start(dryRun ? "Validating publish lock" : "Publishing packages");
	const plan = customPublish ? await customPublish() : await tegami.publish({
		dryRun,
		packages: options.packages && options.packages.length > 0 ? options.packages : void 0
	});
	if (plan === "skipped") {
		s.stop(dryRun ? "No publish lock to validate" : "Nothing to publish");
		outro(`No publishable packages were found in ${context.lockPath}.`);
		return false;
	}
	s.stop(dryRun ? "Publish lock validated" : "Publish complete");
	const lines = [];
	let hasFailed = false;
	for (const [id, packagePlan] of plan.packages) {
		if (!packagePlan.preflight.shouldPublish) continue;
		const result = packagePlan.publishResult;
		const pkg = context.graph.get(id);
		if (result.type === "failed") hasFailed = true;
		let message = `${result.type} ${pkg.id} - ${pkg.version}`;
		const distTag = packagePlan.npm?.distTag;
		if (distTag) message += ` (npm dist-tag: ${distTag})`;
		if (result.type === "failed" && result.error) message += `: ${result.error}`;
		lines.push(message);
	}
	note(lines.join("\n"), dryRun ? "Publish dry run" : "Publish result");
	if (hasFailed) {
		process.exitCode = 1;
		outro("Failed to publish.");
		return false;
	}
	outro(dryRun ? "Publish lock is valid." : "Packages published.");
	return true;
}
async function runInitAgent(tegami, options) {
	intro("Init agent instructions");
	const context = await tegami._internal.context();
	const s = spinner();
	s.start("Writing AGENTS.md");
	const result = await initAgent(context, options);
	s.stop(result.created ? "Created AGENTS.md" : "Appended to AGENTS.md");
	note(path.relative(context.cwd, result.path) || "AGENTS.md", "Agent instructions");
	outro("Agents can follow AGENTS.md to write changelogs.");
}
async function runCleanup(tegami) {
	intro("Cleanup publish lock");
	const s = spinner();
	s.start("Checking publish lock status");
	const result = await tegami.cleanup();
	const { lockPath } = await tegami._internal.context();
	s.stop(result.state === "removed" ? "Publish lock removed" : "Publish lock kept");
	if (result.state === "removed") {
		outro(`Removed ${lockPath}.`);
		return;
	}
	if (result.reason === "no-plan") {
		outro(`No publish lock found at ${lockPath}.`);
		return;
	}
	outro(`Publish lock at ${lockPath} is still pending. Publish it before cleanup.`);
}
//#endregion
export { createCli, runCli };
