// tegami/dist/_assertGuard-cQ6dIdMj.js
//#region ../../node_modules/.pnpm/typia@12.1.1_@types+node@26.1.1_typescript@6.0.3/node_modules/typia/lib/TypeGuardError.mjs
/**
* Error thrown when type assertion fails.
*
* Thrown by {@link assert}, {@link assertGuard}, and other assert-family
* functions when input doesn't match expected type `T`. Contains detailed
* information about the first assertion failure:
*
* - `method`: Which typia function threw (e.g., `"typia.assert"`)
* - `path`: Property path where error occurred (e.g., `"input.user.age"`)
* - `expected`: Expected type string (e.g., `"number & ExclusiveMinimum<19>"`)
* - `value`: Actual value that failed validation
*
* @template T Expected type (for type safety)
*/
var TypeGuardError = class extends Error {
	/**
	* Name of the typia method that threw this error.
	*
	* E.g., `"typia.assert"`, `"typia.assertEquals"`, `"typia.assertGuard"`.
	*/
	method;
	/**
	* Property path where assertion failed.
	*
	* Uses dot notation for nested properties. `undefined` if error occurred at
	* root level.
	*
	* E.g., `"input.age"`, `"input.profile.email"`, `"input[0].name"`.
	*/
	path;
	/**
	* String representation of expected type.
	*
	* E.g., `"string"`, `"number & ExclusiveMinimum<19>"`, `"{ name: string; age:
	* number }"`.
	*/
	expected;
	/**
	* Actual value that failed assertion.
	*
	* The raw value at the error path, useful for debugging.
	*/
	value;
	/**
	* Optional human-readable error description.
	*
	* Primarily for AI agent libraries or custom validation scenarios needing
	* additional context. Standard assertions rely on `path`, `expected`, and
	* `value` for error reporting.
	*/
	description;
	/**
	* Phantom property for TypeScript type safety.
	*
	* Not used at runtime—exists only to preserve generic type `T` in the type
	* system. Always `undefined`.
	*
	* @internal
	*/
	fake_expected_typed_value_;
	/**
	* Creates a new TypeGuardError instance.
	*
	* @param props Error properties
	*/
	constructor(props) {
		super(props.message || `Error on ${props.method}(): invalid type${props.path ? ` on ${props.path}` : ""}, expect to be ${props.expected}`);
		const proto = new.target.prototype;
		if (Object.setPrototypeOf) Object.setPrototypeOf(this, proto);
		else this.__proto__ = proto;
		this.method = props.method;
		this.path = props.path;
		this.expected = props.expected;
		this.value = props.value;
		if (props.description || props.value === void 0) this.description = props.description ?? [
			"The value at this path is `undefined`.",
			"",
			`Please fill the \`${props.expected}\` typed value next time.`
		].join("\n");
	}
};
//#endregion
//#region ../../node_modules/.pnpm/typia@12.1.1_@types+node@26.1.1_typescript@6.0.3/node_modules/typia/lib/internal/_assertGuard.mjs
const _assertGuard = (exceptionable, props, factory) => {
	if (exceptionable === true) if (factory) throw factory(props);
	else throw new TypeGuardError(props);
	return false;
};
//#endregion
export { _assertGuard as t };
