// tegami/dist/_createStandardSchema-DFfVlh8x.js
//#region ../../node_modules/.pnpm/typia@12.1.1_@types+node@26.1.1_typescript@6.0.3/node_modules/typia/lib/internal/_validateReport.mjs
const _validateReport = (array) => {
	const reportable = (path) => {
		if (array.length === 0) return true;
		const last = array[array.length - 1].path;
		return path.length > last.length || last.substring(0, path.length) !== path;
	};
	return (exceptable, error) => {
		if (exceptable && reportable(error.path)) {
			if (error.value === void 0) error.description ??= [
				"The value at this path is `undefined`.",
				"",
				`Please fill the \`${error.expected}\` typed value next time.`
			].join("\n");
			array.push(error);
		}
		return false;
	};
};
//#endregion
//#region ../../node_modules/.pnpm/typia@12.1.1_@types+node@26.1.1_typescript@6.0.3/node_modules/typia/lib/internal/_createStandardSchema.mjs
const _createStandardSchema = (fn) => Object.assign(fn, { "~standard": {
	version: 1,
	vendor: "typia",
	validate: (input) => {
		const result = fn(input);
		if (result.success) return { value: result.data };
		else return { issues: result.errors.map((error) => ({
			message: `expected ${error.expected}, got ${error.value}`,
			path: typiaPathToStandardSchemaPath(error.path)
		})) };
	}
} });
var PathParserState;
(function(PathParserState) {
	PathParserState[PathParserState["Start"] = 0] = "Start";
	PathParserState[PathParserState["Property"] = 1] = "Property";
	PathParserState[PathParserState["StringKey"] = 2] = "StringKey";
	PathParserState[PathParserState["NumberKey"] = 3] = "NumberKey";
})(PathParserState || (PathParserState = {}));
const typiaPathToStandardSchemaPath = (path) => {
	if (!path.startsWith("$input")) throw new Error(`Invalid path: ${JSON.stringify(path)}`);
	const segments = [];
	let currentSegment = "";
	let state = PathParserState.Start;
	let index = 5;
	while (index < path.length - 1) {
		index++;
		const char = path[index];
		if (state === PathParserState.Property) if (char === "." || char === "[") {
			segments.push({ key: currentSegment });
			state = PathParserState.Start;
		} else if (index === path.length - 1) {
			currentSegment += char;
			segments.push({ key: currentSegment });
			index++;
			state = PathParserState.Start;
		} else currentSegment += char;
		else if (state === PathParserState.StringKey) if (char === "\"") {
			segments.push({ key: JSON.parse(currentSegment + char) });
			index += 2;
			state = PathParserState.Start;
		} else if (char === "\\") {
			currentSegment += path[index];
			index++;
			currentSegment += path[index];
		} else currentSegment += char;
		else if (state === PathParserState.NumberKey) if (char === "]") {
			segments.push({ key: Number.parseInt(currentSegment) });
			index++;
			state = PathParserState.Start;
		} else currentSegment += char;
		if (state === PathParserState.Start && index < path.length - 1) {
			const newChar = path[index];
			currentSegment = "";
			if (newChar === "[") if (path[index + 1] === "\"") {
				state = PathParserState.StringKey;
				index++;
				currentSegment = "\"";
			} else state = PathParserState.NumberKey;
			else if (newChar === ".") state = PathParserState.Property;
			else throw new Error("Unreachable: pointer points invalid character");
		}
	}
	if (state !== PathParserState.Start) throw new Error(`Failed to parse path: ${JSON.stringify(path)}`);
	return segments;
};
//#endregion
export { _validateReport as n, _createStandardSchema as t };
