// tegami/dist/error-CAsrGb6e.js
//#region src/utils/common.ts
const isCI = () => Boolean(process.env.CI);
async function somePromise(promises, fn) {
	return new Promise((res, reject) => {
		const n = promises.length;
		if (n === 0) res(false);
		let i = 0;
		for (const promise of promises) {
			if (promise instanceof Promise) {
				promise.then((v) => {
					if (fn(v)) return res(true);
					i++;
					if (i === n) res(false);
				}).catch(reject);
				continue;
			}
			if (fn(promise)) return res(true);
			i++;
			if (i === n) res(false);
		}
	});
}
function joinPath(...paths) {
	let out = "";
	for (const path of paths) {
		if (path.length === 0) continue;
		if (!out) {
			out = path;
			continue;
		}
		if (!out.endsWith("/") && !path.startsWith("/")) out += "/";
		out += path;
	}
	return out;
}
function cached(cacheKey, fn, cacheMap = /* @__PURE__ */ new Map()) {
	return (...args) => {
		const key = cacheKey(...args);
		let out = cacheMap.get(key);
		if (!out) {
			out = fn(...args);
			cacheMap.set(key, out);
		}
		return out;
	};
}
//#endregion
//#region src/utils/error.ts
var CancelledError = class extends Error {
	constructor() {
		super("Cancelled.");
	}
};
function execFailure(context, result) {
	const lines = [context, `(exit ${result.exitCode})`];
	const out = result.stdout.trim();
	const err = result.stderr.trim();
	if (out) lines.push(out);
	if (err) lines.push(err);
	return new Error(redactSensitiveTokens(lines.join("\n")));
}
async function fetchFailure(context, response) {
	const body = (await response.text()).trim();
	const message = body ? `${context}: ${body}` : context;
	return new Error(redactSensitiveTokens(message));
}
function isNodeError(error) {
	return error instanceof Error && "code" in error;
}
const SENSITIVE_TOKEN_PATTERNS = [
	/\b(npm_[a-zA-Z0-9]{36,})\b/g,
	/\b(gh[pousrsa]_[A-Za-z0-9_]{36,})\b/g,
	/\b(bearer\s+[a-z0-9\-_]{20,})\b/gi,
	/\b(glpat-[A-Za-z0-9-_]{20,})\b/g
];
/**
* should not be needed, but we cannot be sure if the used CLI tools will expose secrets by accident.
*/
function redactSensitiveTokens(text) {
	if (!isCI()) return text;
	let redacted = text;
	for (const pattern of SENSITIVE_TOKEN_PATTERNS) redacted = redacted.replace(pattern, "[REDACTED_TOKEN]");
	return redacted;
}
async function handlePluginError(plugin, hookName, callback) {
	try {
		return await callback();
	} catch (error) {
		const details = error instanceof Error ? error.message : String(error);
		throw new Error(`Plugin "${plugin.name}" failed during ${hookName}:\n${redactSensitiveTokens(details)}`, { cause: error });
	}
}
//#endregion
export { isNodeError as a, joinPath as c, handlePluginError as i, somePromise as l, execFailure as n, cached as o, fetchFailure as r, isCI as s, CancelledError as t };
