// tegami/dist/generate-Cip5o2xy.js
import { o as maxBump, t as bumpDepth } from "./semver-DrtaCCZK.js";
import { n as execFailure } from "./error-CAsrGb6e.js";
import { n as renderChangelog } from "./shared-BraxbVVG.js";
import { x } from "tinyexec";
//#region src/utils/conventional-commit.ts
/**
* Conventional commit header per conventionalcommits.org and semantic-release defaults.
* @see https://www.conventionalcommits.org/en/v1.0.0/
*/
const CONVENTIONAL_COMMIT_HEADER = /^(?<type>\w+)(?:\((?<scope>[^)]*)\))?(?<breaking>!)?: (?<title>.+)$/;
const BREAKING_CHANGE_FOOTER = /^BREAKING[ -]CHANGE:/m;
/** Parse conventional commits and resolve scopes against the workspace graph. */
function createConventionalCommitParser(graph) {
	const byShortName = /* @__PURE__ */ new Map();
	for (const pkg of graph.getPackages()) {
		const slash = pkg.name.lastIndexOf("/");
		const short = slash >= 0 ? pkg.name.slice(slash + 1) : pkg.name;
		const names = byShortName.get(short);
		if (names) names.push(pkg.name);
		else byShortName.set(short, [pkg.name]);
	}
	function resolvePackages(scope) {
		if (!scope) return [];
		const packages = /* @__PURE__ */ new Set();
		for (const item of scope.split(",")) {
			const name = item.trim();
			if (!name) continue;
			const direct = graph.getByName(name);
			if (direct.length > 0) {
				for (const pkg of direct) packages.add(pkg.name);
				continue;
			}
			const byShort = byShortName.get(name);
			if (byShort) {
				for (const pkgName of byShort) packages.add(pkgName);
				continue;
			}
			packages.add(name);
		}
		return Array.from(packages);
	}
	return function parseConventionalCommit(subject, body = "") {
		const match = CONVENTIONAL_COMMIT_HEADER.exec(subject.trim());
		if (!match?.groups) return;
		const { type, scope, breaking, title } = match.groups;
		if (!type || !title) return;
		const trimmedTitle = title.trim();
		if (!trimmedTitle) return;
		const trimmedScope = scope?.trim();
		return {
			type: type.toLowerCase(),
			packages: resolvePackages(trimmedScope || void 0),
			breaking: Boolean(breaking) || BREAKING_CHANGE_FOOTER.test(body),
			title: trimmedTitle
		};
	};
}
/** Map releasable conventional commit types to semver bumps (semantic-release defaults). */
function conventionalCommitToBump(type, breaking) {
	if (breaking) return "major";
	switch (type) {
		case "feat": return "minor";
		case "fix":
		case "perf":
		case "revert": return "patch";
		default: return;
	}
}
//#endregion
//#region src/changelog/generate.ts
async function generateFromCommits(context, { from, to } = {}) {
	const commits = await readConventionalCommits(context, from, to);
	const groups = /* @__PURE__ */ new Map();
	for (const commit of commits) {
		const key = commit.packages.join("\0");
		const group = groups.get(key);
		if (group) group.push(commit);
		else groups.set(key, [commit]);
	}
	return Array.from(groups, ([key, changes], index) => {
		const packageNames = key ? key.split("\0") : [];
		let bumpType = "patch";
		for (const change of changes) bumpType = maxBump(change.type, bumpType);
		const packageBumpMap = Object.fromEntries(packageNames.map((name) => [name, bumpType]));
		const content = renderChangelog({ packages: packageBumpMap }, changes.map((change) => {
			const heading = "#".repeat(bumpDepth(change.type));
			if (!change.body) return `${heading} ${change.title}`;
			return `${heading} ${change.title}\n\n${change.body}`;
		}).join("\n\n"));
		return {
			filename: changelogFilename(index),
			content,
			packages: packageBumpMap,
			changes
		};
	});
}
async function readConventionalCommits(context, from, to) {
	from ??= await latestTag(context.cwd);
	to ??= "HEAD";
	const args = [
		"log",
		"--no-merges",
		"--format=%H%x1f%s%x1f%b%x1e"
	];
	if (from) args.push(`${from}..${to}`);
	else if (to !== "HEAD") args.push(to);
	const result = await x("git", args, { nodeOptions: { cwd: context.cwd } });
	if (result.exitCode !== 0) throw execFailure(`Unable to read git commits from ${from ?? "start"} to ${to}`, result);
	const parseCommit = createConventionalCommitParser(context.graph);
	const changes = [];
	for (const record of result.stdout.split("")) {
		const [hash, subject, body = ""] = record.replace(/^\n+|\n+$/g, "").split("");
		if (!hash || !subject) continue;
		const parsed = parseCommit(subject, body);
		if (!parsed) continue;
		const bump = conventionalCommitToBump(parsed.type, parsed.breaking);
		if (!bump) continue;
		changes.push({
			hash,
			subject,
			body: body.trim(),
			packages: parsed.packages,
			type: bump,
			title: parsed.title.charAt(0).toUpperCase() + parsed.title.slice(1)
		});
	}
	return changes;
}
async function latestTag(cwd) {
	const result = await x("git", [
		"describe",
		"--tags",
		"--abbrev=0"
	], { nodeOptions: { cwd } });
	if (result.exitCode !== 0) return;
	return result.stdout.trim() || void 0;
}
function changelogFilename(disambiguator = 0) {
	const date = /* @__PURE__ */ new Date();
	return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}-${String(date.getDate()).padStart(2, "0")}-${(Date.now() + disambiguator).toString(36)}.md`;
}
//#endregion
export { generateFromCommits as n, changelogFilename as t };
