// tegami/dist/graph-DaJ28Y4f.js
import { n as bumpVersion } from "./semver-DrtaCCZK.js";
//#region src/graph.ts
/** Package discovered in the workspace. */
var WorkspacePackage = class {
	/** note: this will only be available after package graph is resolved */
	group;
	/** note: this will only be available after package graph is resolved */
	options = {};
	get id() {
		return `${this.manager}:${this.name}`;
	}
	/** create the initial draft. */
	initDraft() {
		return { bumpVersion(pkg) {
			if (!pkg.version) return;
			return bumpVersion(pkg.version, this.type, this.prerelease);
		} };
	}
	/** configure an initial draft to match script-level configs. */
	configureDraft({ draft }) {
		const { prerelease = this.group?.options?.prerelease } = this.options;
		if (prerelease !== void 0) draft.prerelease = prerelease;
	}
};
/**
* Unified graph for discovered workspace packages.
*
* This is only used as a storage for all indexed packages.
* For registry-specific relationships (e.g. virtual workspaces), they are stored in the provider plugin internally.
*/
var PackageGraph = class {
	packages = /* @__PURE__ */ new Map();
	groups = /* @__PURE__ */ new Map();
	constructor(packages = []) {
		for (const pkg of packages) this.add(pkg);
	}
	getPackages() {
		return Array.from(this.packages.values());
	}
	/** Get a package by exact id. */
	get(id) {
		return this.packages.get(id);
	}
	/** Get packages by id, `group:name`, or every package matching a name. */
	getByName(nameOrId) {
		const exact = this.packages.get(nameOrId);
		if (exact) return [exact];
		if (nameOrId.startsWith("group:")) return this.getGroup(nameOrId.slice(6))?.packages ?? [];
		const out = [];
		for (const value of this.packages.values()) if (value.name === nameOrId) out.push(value);
		return out;
	}
	/** scan package into graph, if the package id already exists, replace the existing one in graph */
	add(pkg) {
		this.delete(pkg.id);
		this.packages.set(pkg.id, pkg);
	}
	delete(id) {
		this.packages.delete(id);
		for (const group of this.groups.values()) {
			const index = group.packages.findIndex((pkg) => pkg.id === id);
			if (index >= 0) group.packages.splice(index, 1);
		}
	}
	getPackageGroup(pkgId) {
		return this.packages.get(pkgId)?.group;
	}
	getGroups() {
		return Array.from(this.groups.values());
	}
	getGroup(name) {
		return this.groups.get(name);
	}
	registerGroup(name, options) {
		const existing = this.groups.get(name);
		if (existing) {
			existing.options = options;
			return existing;
		}
		const group = {
			name,
			options,
			packages: []
		};
		this.groups.set(name, group);
		return group;
	}
	addGroupMember(groupId, id) {
		const group = this.groups.get(groupId);
		const pkg = this.packages.get(id);
		if (!group || !pkg || pkg.group) return;
		pkg.group = group;
		group.packages.push(pkg);
	}
	removeGroupMember(group, id) {
		const entry = this.groups.get(group);
		const pkg = this.packages.get(id);
		if (!entry || !pkg || pkg.group !== entry) return;
		const index = entry.packages.findIndex((pkg) => pkg.id === id);
		if (index >= 0) entry.packages.splice(index, 1);
		pkg.group = void 0;
	}
	unregisterGroup(name) {
		const group = this.groups.get(name);
		if (!group) return;
		for (const pkg of group.packages) {
			const entry = this.packages.get(pkg.id);
			if (entry?.group === group) entry.group = void 0;
		}
		this.groups.delete(name);
	}
};
//#endregion
export { WorkspacePackage as n, PackageGraph as t };
