// tegami/dist/index.js
import { n as generateFromCommits } from "./generate-Cip5o2xy.js";
import { i as handlePluginError } from "./error-CAsrGb6e.js";
import { t as npm } from "./npm-BHUzBLSi.js";
import { a as createDraft, c as parseChangelogFile, i as runPublishPlan, l as readChangelogEntries, n as publishPlanStatus, r as runPreflights, t as initPublishPlan } from "./publish-CQifflae.js";
import { n as WorkspacePackage, t as PackageGraph } from "./graph-DaJ28Y4f.js";
import { mkdir, rm, writeFile } from "node:fs/promises";
import path, { join } from "node:path";
//#region src/context.ts
async function createTegamiContext(options = {}) {
	const cwd = options.cwd ? path.resolve(options.cwd) : process.cwd();
	const changelogDir = path.resolve(cwd, options.changelogDir ?? ".tegami");
	const ctx = {
		cwd,
		changelogDir,
		lockPath: options.lockPath ? path.resolve(cwd, options.lockPath) : path.join(changelogDir, "publish-lock.yaml"),
		options,
		plugins: resolvePlugins([npm(options.npm), ...options.plugins ?? []]),
		graph: new PackageGraph()
	};
	for (const plugin of ctx.plugins) await handlePluginError(plugin, "init", () => plugin.init?.call(ctx));
	return ctx;
}
async function resolveGraph(ctx) {
	const { options, graph } = ctx;
	for (const plugin of ctx.plugins) await handlePluginError(plugin, "resolve", () => plugin.resolve?.call(ctx));
	const ignoreMatchers = options.ignore?.map((pattern) => {
		if (pattern instanceof RegExp) return (pkg) => pattern.test(pkg.name) || pattern.test(pkg.id);
		return (pkg) => pkg.name === pattern || pkg.id === pattern;
	});
	for (const [name, groupOptions] of Object.entries(options.groups ?? {})) graph.registerGroup(name, groupOptions);
	let getPackageOptions;
	if (typeof options.packages === "function") getPackageOptions = options.packages;
	else if (options.packages) {
		const packages = options.packages;
		getPackageOptions = (pkg) => packages[pkg.id] ?? packages[pkg.name];
	}
	for (const pkg of graph.getPackages()) {
		if (ignoreMatchers && ignoreMatchers.some((matcher) => matcher(pkg))) {
			graph.delete(pkg.id);
			continue;
		}
		const packageOptions = getPackageOptions?.(pkg);
		if (!packageOptions) continue;
		pkg.options = packageOptions;
		if (packageOptions.group) graph.addGroupMember(packageOptions.group, pkg.id);
	}
}
const PLUGIN_ORDER = {
	pre: 0,
	default: 1,
	post: 2
};
function resolvePlugins(plugins = []) {
	return plugins.flat(Infinity).sort((a, b) => PLUGIN_ORDER[a.enforce ?? "default"] - PLUGIN_ORDER[b.enforce ?? "default"]);
}
//#endregion
//#region src/index.ts
/** Create a Tegami project handle. */
function tegami(options = {}) {
	let $context;
	let $contextResolved;
	function getContext() {
		return $context ??= createTegamiContext(options);
	}
	function getContextResolved() {
		return $contextResolved ??= getContext().then(async (ctx) => {
			await resolveGraph(ctx);
			return ctx;
		});
	}
	return {
		async generateChangelog(generateOptions = {}) {
			const { write = true } = generateOptions;
			const context = await getContextResolved();
			const changelogs = await generateFromCommits(context, generateOptions);
			if (write && changelogs.length > 0) {
				await mkdir(context.changelogDir, { recursive: true });
				await Promise.all(changelogs.map((entry) => writeFile(join(context.changelogDir, entry.filename), entry.content)));
			}
			return changelogs;
		},
		_internal: {
			options,
			context: getContextResolved,
			contextUnresolved: getContext
		},
		async draft() {
			const context = await getContextResolved();
			const changelogs = await readChangelogEntries(context);
			if (context.options.conventionalCommits) {
				const generated = await generateFromCommits(context);
				for (const entry of generated) {
					const parsed = parseChangelogFile(entry.filename, entry.content);
					if (!parsed) continue;
					parsed.virtual = true;
					changelogs.push(parsed);
				}
			}
			return createDraft(changelogs, context);
		},
		async getPublishStatus(publishOptions = {}) {
			const context = await getContextResolved();
			const plan = await initPublishPlan(context, publishOptions);
			if (!plan) return { status: "none" };
			await runPreflights(context, plan);
			return publishPlanStatus(plan, context);
		},
		async publishStatus(publishOptions = {}) {
			const { status } = await this.getPublishStatus(publishOptions);
			return status === "none" ? "idle" : status;
		},
		async publish(publishOptions = {}) {
			const context = await getContextResolved();
			const plan = await initPublishPlan(context, publishOptions);
			if (!plan) return "skipped";
			await runPreflights(context, plan);
			if ((await publishPlanStatus(plan, context)).status === "success") return "skipped";
			await runPublishPlan(context, plan);
			if (Array.from(plan.packages.values()).every((pkg) => pkg.publishResult.type === "skipped")) return "skipped";
			return plan;
		},
		async cleanup(publishOptions = {}) {
			const context = await getContextResolved();
			const plan = await initPublishPlan(context, publishOptions);
			if (!plan) return {
				state: "skipped",
				reason: "no-plan"
			};
			await runPreflights(context, plan);
			const { status } = await publishPlanStatus(plan, context);
			if (status !== "success") return {
				state: "skipped",
				reason: "pending"
			};
			await rm(context.lockPath, { force: true });
			return { state: "removed" };
		}
	};
}
//#endregion
export { PackageGraph, WorkspacePackage, tegami };
