// tegami/dist/publish-CQifflae.js
import { o as maxBump } from "./semver-DrtaCCZK.js";
import { i as handlePluginError, l as somePromise } from "./error-CAsrGb6e.js";
import { t as _accessExpressionAsString } from "./_accessExpressionAsString-DAz6bHI_.js";
import { n as _validateReport, t as _createStandardSchema } from "./_createStandardSchema-DFfVlh8x.js";
import { n as renderChangelog, r as validateChangelogFrontmatter, t as getPackageBumps } from "./shared-BraxbVVG.js";
import { simpleGenerator } from "./generators/simple.js";
import { t as _assertGuard } from "./_assertGuard-cQ6dIdMj.js";
import fs, { mkdir, readFile, readdir, rm, writeFile } from "node:fs/promises";
import path, { basename, dirname, join } from "node:path";
import * as semver$1 from "semver";
import { parse as parse$1, stringify } from "yaml";
//#region src/utils/frontmatter.ts
/**
* Inspired by https://github.com/jonschlinkert/gray-matter
*/
const regex = /^---\r?\n(.+?)\r?\n---\r?\n?/s;
/**
* parse frontmatter, it supports only yaml format
*/
function frontmatter(input) {
	const output = {
		matter: "",
		data: {},
		content: input
	};
	const match = regex.exec(input);
	if (!match) return output;
	output.matter = match[0];
	output.content = input.slice(match[0].length);
	output.data = parse$1(match[1]) ?? {};
	return output;
}
//#endregion
//#region src/changelog/parse.ts
async function getChangelogFiles(context) {
	return (await readdir(context.changelogDir).catch(() => [])).filter((file) => file.endsWith(".md"));
}
async function readChangelogEntries(context) {
	const dir = context.changelogDir;
	const files = await getChangelogFiles(context);
	return (await Promise.all(files.map(async (file) => {
		const filePath = join(dir, file);
		const content = await readFile(filePath, "utf8");
		return parseChangelogFile(basename(filePath), content);
	}))).filter((v) => v !== void 0);
}
/** Parse one changelog markdown file into release entries. */
function parseChangelogFile(filename, content) {
	const parsed = frontmatter(content);
	const validated = validateChangelogFrontmatter(parsed.data);
	if (!validated.success) return;
	const { packages: packagesConfig } = validated.data;
	if (!packagesConfig) return;
	const data = validated.data;
	let headingBump;
	const packages = /* @__PURE__ */ new Map();
	const sections = [];
	for (const section of parseMarkdownSections(parsed.content)) {
		const sectionBumpType = headingToBump(section.depth);
		if (sectionBumpType) headingBump = headingBump ? maxBump(headingBump, sectionBumpType) : sectionBumpType;
		sections.push({
			depth: section.depth,
			title: section.title,
			content: section.content
		});
	}
	if (sections.length === 0) return;
	if (Array.isArray(packagesConfig)) {
		if (!headingBump) return;
		for (const pkg of packagesConfig) packages.set(pkg, { type: headingBump });
	} else for (const [k, v] of Object.entries(packagesConfig)) {
		let config;
		if (typeof v === "string") config = { type: v };
		else if (v === null) config = { type: headingBump };
		else config = v;
		if (config.type || config.replay?.length) packages.set(k, config);
	}
	if (packages.size === 0) return;
	const entry = {
		id: filename,
		filename,
		subject: data.subject,
		packages,
		sections,
		_raw_body: parsed.content,
		getRawContent() {
			return renderChangelog({
				subject: this.subject,
				packages: Object.fromEntries(this.packages.entries())
			}, entry._raw_body);
		}
	};
	return entry;
}
function formatReplayCondition(condition) {
	switch (condition.on) {
		case "exit-prerelease": return `exit-prerelease(${condition.name})`;
		case "enter-prerelease": return `prerelease(${condition.name})`;
		case "version": return `${condition.name}@${condition.version}`;
	}
}
function parseReplayCondition(condition) {
	let match;
	if (match = /^exit-prerelease\((.+)\)$/.exec(condition)) return {
		on: "exit-prerelease",
		name: match[1]
	};
	if (match = /^prerelease\((.+)\)$/.exec(condition)) return {
		on: "enter-prerelease",
		name: match[1]
	};
	if (match = /^exit prerelease:\s*(.+)$/.exec(condition)) return {
		on: "exit-prerelease",
		name: match[1]
	};
	if (match = /^(.+)@([^@]+)$/.exec(condition)) return {
		on: "version",
		name: match[1],
		version: match[2]
	};
	return null;
}
function parseMarkdownSections(markdown) {
	const sections = [];
	const lines = markdown.split(/\r\n|\r|\n/);
	let current;
	let fence;
	for (const line of lines) {
		const fenceMarker = line.match(/^ {0,3}(`{3,}|~{3,})/);
		if (fenceMarker) {
			const marker = fenceMarker[1];
			if (!fence) fence = marker;
			else if (marker[0] === fence[0] && marker.length >= fence.length) fence = void 0;
		}
		if (!fence) {
			const heading = parseHeading(line);
			if (heading) {
				if (current) sections.push(toMarkdownSection(current));
				current = {
					...heading,
					contentLines: []
				};
				continue;
			}
		}
		current?.contentLines.push(line);
	}
	if (current) sections.push(toMarkdownSection(current));
	return sections;
}
function parseHeading(line) {
	const match = line.match(/^ {0,3}(#{1,6})(?:[ \t]+|$)(.*)$/);
	if (!match) return;
	return {
		depth: match[1].length,
		title: stripClosingHeadingSequence(match[2]).trim()
	};
}
function stripClosingHeadingSequence(value) {
	return value.replace(/[ \t]+#{1,}[ \t]*$/, "");
}
function toMarkdownSection(section) {
	return {
		depth: section.depth,
		title: section.title,
		content: section.contentLines.join("\n").trim()
	};
}
function headingToBump(depth) {
	switch (depth) {
		case 1: return "major";
		case 2: return "minor";
		case 3: return "patch";
	}
}
//#endregion
//#region src/plans/lock.ts
/**
* the data structure of `publish-lock.yaml` file.
*/
var PublishLock = class PublishLock {
	/** namespace -> data array */
	data;
	constructor(input = /* @__PURE__ */ new Map()) {
		if (input instanceof PublishLock) {
			this.data = /* @__PURE__ */ new Map();
			for (const [k, v] of input.data) this.data.set(k, [...v]);
		} else this.data = input;
	}
	/** write data to namespace, note that the `data` must be serializable in yaml */
	write(namespace, data) {
		let arr = this.data.get(namespace);
		if (!arr) {
			arr = [];
			this.data.set(namespace, arr);
		}
		arr.push(data);
	}
	read(namespace) {
		return this.data.get(namespace)?.shift();
	}
	size(namespace) {
		return this.data.get(namespace)?.length ?? 0;
	}
	serialize() {
		let res = stringify(this.data, {
			sortMapEntries: true,
			lineWidth: 0,
			blockQuote: false,
			doubleQuotedAsJSON: true
		});
		if (!res.endsWith("\n")) res += "\n";
		return res;
	}
};
const assertPublishLockData = (() => {
	const _io0 = (input) => Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return Array.isArray(value);
	});
	const _ao0 = (input, _path, _exceptionable = true) => false === _exceptionable || Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return Array.isArray(value) || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "Array<unknown>",
			value
		}, _errorFactory);
	});
	const __is = (input) => "object" === typeof input && null !== input && false === Array.isArray(input) && _io0(input);
	let _errorFactory;
	return (input, errorFactory) => {
		if (false === __is(input)) {
			_errorFactory = errorFactory;
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input && false === Array.isArray(input) || _assertGuard(true, {
				method: "typia.createAssert",
				path: _path + "",
				expected: "Record<string, Array<unknown>>",
				value: input
			}, _errorFactory)) && _ao0(input, _path + "", true) || _assertGuard(true, {
				method: "typia.createAssert",
				path: _path + "",
				expected: "Record<string, Array<unknown>>",
				value: input
			}, _errorFactory))(input, "$input", true);
		}
		return input;
	};
})();
function parsePublishLock(content) {
	const data = assertPublishLockData(parse$1(content));
	return new PublishLock(new Map(Object.entries(data)));
}
//#endregion
//#region src/plans/policy.ts
function groupPolicy(_ctx) {
	return {
		id: "group",
		onUpdate({ pkg, packageDraft }) {
			if (!packageDraft.type) return;
			const group = pkg.group;
			if (!group || !group.options.syncBump) return;
			for (const member of group.packages) {
				if (member === pkg) continue;
				this.bumpPackage(member, {
					type: packageDraft.type,
					reason: `sync "${group.name}" group package versions`
				});
			}
		}
	};
}
//#endregion
//#region src/plans/draft.ts
const validateChangelogStore = (() => {
	const _io0 = (input) => "0.0.0" === input.v && "string" === typeof input.filename && "string" === typeof input.content;
	const _vo0 = (input, _path, _exceptionable = true) => [
		"0.0.0" === input.v || _report(_exceptionable, {
			path: _path + ".v",
			expected: "\"0.0.0\"",
			value: input.v
		}),
		"string" === typeof input.filename || _report(_exceptionable, {
			path: _path + ".filename",
			expected: "string",
			value: input.filename
		}),
		"string" === typeof input.content || _report(_exceptionable, {
			path: _path + ".content",
			expected: "string",
			value: input.content
		})
	].every((flag) => flag);
	const __is = (input) => "object" === typeof input && null !== input && _io0(input);
	let errors;
	let _report;
	return _createStandardSchema((input) => {
		if (false === __is(input)) {
			errors = [];
			_report = _validateReport(errors);
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input || _report(true, {
				path: _path + "",
				expected: "ChangelogStore",
				value: input
			})) && _vo0(input, _path + "", true) || _report(true, {
				path: _path + "",
				expected: "ChangelogStore",
				value: input
			}))(input, "$input", true);
			const success = 0 === errors.length;
			return success ? {
				success,
				data: input
			} : {
				success,
				errors,
				data: input
			};
		}
		return {
			success: true,
			data: input
		};
	});
})();
const validatePackageStore = (() => {
	const _io0 = (input) => "string" === typeof input.id && "boolean" === typeof input.updated && (void 0 === input.changelogIds || Array.isArray(input.changelogIds) && input.changelogIds.every((elem) => "string" === typeof elem));
	const _vo0 = (input, _path, _exceptionable = true) => [
		"string" === typeof input.id || _report(_exceptionable, {
			path: _path + ".id",
			expected: "string",
			value: input.id
		}),
		"boolean" === typeof input.updated || _report(_exceptionable, {
			path: _path + ".updated",
			expected: "boolean",
			value: input.updated
		}),
		void 0 === input.changelogIds || (Array.isArray(input.changelogIds) || _report(_exceptionable, {
			path: _path + ".changelogIds",
			expected: "(Array<string> | undefined)",
			value: input.changelogIds
		})) && input.changelogIds.map((elem, _index2) => "string" === typeof elem || _report(_exceptionable, {
			path: _path + ".changelogIds[" + _index2 + "]",
			expected: "string",
			value: elem
		})).every((flag) => flag) || _report(_exceptionable, {
			path: _path + ".changelogIds",
			expected: "(Array<string> | undefined)",
			value: input.changelogIds
		})
	].every((flag) => flag);
	const __is = (input) => "object" === typeof input && null !== input && _io0(input);
	let errors;
	let _report;
	return _createStandardSchema((input) => {
		if (false === __is(input)) {
			errors = [];
			_report = _validateReport(errors);
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input || _report(true, {
				path: _path + "",
				expected: "PackageStore",
				value: input
			})) && _vo0(input, _path + "", true) || _report(true, {
				path: _path + "",
				expected: "PackageStore",
				value: input
			}))(input, "$input", true);
			const success = 0 === errors.length;
			return success ? {
				success,
				data: input
			} : {
				success,
				errors,
				data: input
			};
		}
		return {
			success: true,
			data: input
		};
	});
})();
/** a draft describes all operations to perform before the actual publishing, such as version bumps. */
var Draft = class {
	context;
	#status = 0;
	/** package id -> draft */
	packages = /* @__PURE__ */ new Map();
	/** id -> changelog */
	changelogs = /* @__PURE__ */ new Map();
	policies = [];
	constructor(context) {
		this.context = context;
		this.policies.push(groupPolicy(context));
	}
	getPackageDrafts() {
		return this.packages;
	}
	getPackageDraft(id) {
		return this.packages.get(id);
	}
	bumpPackage(pkg, { type, reason }) {
		return this.dispatchPackage(pkg, (draft) => {
			draft.type = draft.type ? maxBump(draft.type, type) : type;
			if (reason) {
				draft.bumpReasons ??= /* @__PURE__ */ new Set();
				draft.bumpReasons.add(reason);
			}
		});
	}
	dispatchPackage(pkg, dispatch, onUpdate) {
		const packageDraft = this.getOrInitPackage(pkg);
		const prevVersion = packageDraft.bumpVersion(pkg);
		dispatch(packageDraft);
		if (prevVersion !== packageDraft.bumpVersion(pkg)) {
			onUpdate?.(packageDraft);
			for (const policy of this.policies) policy.onUpdate?.call(this, {
				packageDraft,
				pkg
			});
		}
		return packageDraft;
	}
	getOrInitPackage(pkg) {
		const existing = this.packages.get(pkg.id);
		if (existing) return existing;
		this.packages.set(pkg.id, pkg.initDraft());
		return this.dispatchPackage(pkg, (draft) => pkg.configureDraft({ draft }), (draft) => {
			(draft.bumpReasons ??= /* @__PURE__ */ new Set()).add("align with script-level configs");
		});
	}
	hasPending() {
		const { graph } = this.context;
		for (const [id, draft] of this.packages) {
			const pkg = graph.get(id);
			if (pkg && draft.bumpVersion(pkg) !== pkg.version) return true;
		}
		return false;
	}
	/** get all changelogs, note that this includes replay-only changelogs, as long as they are in the `.tegami` folder. */
	getChangelogs() {
		return Array.from(this.changelogs.values());
	}
	getChangelog(id) {
		return this.changelogs.get(id);
	}
	addChangelog(entry) {
		this.changelogs.set(entry.id, entry);
		for (const [pkg, bumpType] of getPackageBumps(this.context.graph, entry)) attachChangelog(this.bumpPackage(pkg, { type: bumpType }), entry);
	}
	deleteChangelog(id) {
		return this.changelogs.delete(id);
	}
	/** Apply version bumps, lock file, and changelog files. */
	async apply() {
		switch (this.#status) {
			case 1: throw new Error("This draft has already been applied.");
			case 2: throw new Error("There is already a previous apply() run.");
			case 3: throw new Error("The previous apply() run failed, please clear your local git changes and try again.");
		}
		const { graph } = this.context;
		try {
			const snapshots = /* @__PURE__ */ new Map();
			for (const pkg of graph.getPackages()) snapshots.set(pkg.id, { version: pkg.version });
			for (const plugin of this.context.plugins) await handlePluginError(plugin, "applyDraft", () => plugin.applyDraft?.call(this.context, this));
			const updatedChangelogs = this.applyReplays(snapshots);
			const writes = [];
			for (const [id, packageDraft] of this.packages) {
				const pkg = graph.get(id);
				if (!pkg) continue;
				writes.push(this.appendChangelog(pkg, packageDraft));
			}
			for (const entry of this.changelogs.values()) {
				const updated = updatedChangelogs.get(entry.id);
				const filePath = path.join(this.context.changelogDir, entry.filename);
				if (updated) writes.push(mkdir(this.context.changelogDir, { recursive: true }).then(() => writeFile(filePath, updated.getRawContent())));
				else if (!entry.virtual) writes.push(rm(filePath, { force: true }));
			}
			writes.push(this.writeLockFile(snapshots));
			await Promise.all(writes);
			this.#status = 1;
		} catch (e) {
			this.#status = 3;
			throw e;
		}
	}
	/** write persistent data to publish lock */
	async writeLockFile(snapshots) {
		const lock = new PublishLock();
		const changelogs = /* @__PURE__ */ new Set();
		for (const pkg of this.context.graph.getPackages()) {
			const draft = this.getPackageDraft(pkg.id);
			const snapshot = snapshots.get(pkg.id);
			if (!snapshot) continue;
			for (const entry of draft?.changelogs ?? []) changelogs.add(entry);
			lock.write("core:packages", {
				id: pkg.id,
				updated: draft !== void 0 && snapshot.version !== pkg.version,
				changelogIds: draft?.changelogs?.map((entry) => entry.id)
			});
		}
		for (const entry of changelogs) lock.write("core:changelogs", {
			v: "0.0.0",
			filename: entry.filename,
			content: entry.getRawContent()
		});
		for (const plugin of this.context.plugins) await handlePluginError(plugin, "initPublishLock", () => plugin.initPublishLock?.call(this.context, {
			lock,
			draft: this
		}));
		await mkdir(dirname(this.context.lockPath), { recursive: true });
		await writeFile(this.context.lockPath, lock.serialize());
	}
	addPolicy(policy) {
		this.policies.push(policy);
	}
	removePolicy(policy) {
		const idx = this.policies.indexOf(policy);
		if (idx !== -1) this.policies.splice(idx, 1);
	}
	canApply() {
		return this.#status === 0;
	}
	/** Attach replaying changelog entries to packages (already bumped), and return the updated changelog entries. */
	applyReplays(snapshots) {
		const updated = /* @__PURE__ */ new Map();
		const { graph } = this.context;
		const defaultReplays = (name) => {
			const replay = [];
			for (const pkg of graph.getByName(name)) if (this.packages.get(pkg.id)?.prerelease) replay.push(formatReplayCondition({
				on: "exit-prerelease",
				name: pkg.id
			}));
			return replay;
		};
		const isMatch = (condition) => {
			switch (condition.on) {
				case "enter-prerelease": return graph.getByName(condition.name).some((pkg) => {
					const previous = snapshots.get(pkg.id);
					if (!pkg.version || !previous?.version) return false;
					return !semver$1.prerelease(previous.version) && semver$1.prerelease(pkg.version);
				});
				case "exit-prerelease": return graph.getByName(condition.name).some((pkg) => {
					const previous = snapshots.get(pkg.id);
					if (!pkg.version || !previous?.version) return false;
					return semver$1.inc(previous.version, "release") === pkg.version;
				});
				case "version": return graph.getByName(condition.name).some((pkg) => pkg.version === condition.version);
			}
		};
		for (const entry of this.changelogs.values()) {
			const updatedPackages = /* @__PURE__ */ new Map();
			const matchedNames = /* @__PURE__ */ new Set();
			for (const [name, config] of entry.packages) {
				let replay = config.replay;
				if (config.type) replay ??= defaultReplays(name);
				if (!replay || replay.length === 0) continue;
				replay = replay.filter((item) => {
					const condition = parseReplayCondition(item);
					if (condition && isMatch(condition)) {
						matchedNames.add(name);
						return false;
					}
					return true;
				});
				if (replay.length === 0) continue;
				const updatedConfig = {
					...config,
					replay
				};
				delete updatedConfig.type;
				updatedPackages.set(name, updatedConfig);
			}
			if (updatedPackages.size > 0) updated.set(entry.id, {
				...entry,
				packages: updatedPackages
			});
			for (const name of matchedNames) for (const pkg of graph.getByName(name)) attachChangelog(this.getOrInitPackage(pkg), entry);
		}
		return updated;
	}
	async appendChangelog(pkg, draft) {
		if (!draft.changelogs || draft.changelogs.length === 0) return;
		const { generator = simpleGenerator() } = this.context.options;
		const generated = await generator.generate.call(this.context, {
			pkg,
			packageDraft: draft,
			draft: this
		});
		const path = join(pkg.path, "CHANGELOG.md");
		const existing = await readFile(path, "utf8").catch(() => "");
		await writeFile(path, `${generated.trim()}\n\n${existing}`.trimEnd() + "\n");
	}
	/** {@link apply} but for `await using` syntax */
	async [Symbol.asyncDispose]() {
		return this.apply();
	}
};
async function createDraft(changelogs, context) {
	let draft = new Draft(context);
	for (const plugin of context.plugins) {
		const result = await handlePluginError(plugin, "initDraft", () => plugin.initDraft?.call(context, draft));
		if (result) draft = result;
	}
	for (const pkg of context.graph.getPackages()) draft.getOrInitPackage(pkg);
	for (const entry of changelogs) draft.addChangelog(entry);
	return draft;
}
function attachChangelog(draft, entry) {
	if (draft.changelogs?.some((item) => item.id === entry.id)) return;
	draft.changelogs ??= [];
	draft.changelogs.push(entry);
}
//#endregion
//#region src/plans/publish.ts
async function initPublishPlan(context, options) {
	let lock;
	try {
		lock = parsePublishLock(await fs.readFile(context.lockPath, "utf8"));
	} catch {
		return;
	}
	let data;
	const packages = /* @__PURE__ */ new Map();
	const changelogs = /* @__PURE__ */ new Map();
	while (data = lock.read("core:changelogs")) {
		const validated = validateChangelogStore(data);
		if (!validated.success) continue;
		const entry = validated.data;
		const parsed = parseChangelogFile(entry.filename, entry.content);
		if (!parsed) continue;
		changelogs.set(parsed.id, parsed);
	}
	while (data = lock.read("core:packages")) {
		const validated = validatePackageStore(data);
		if (!validated.success) continue;
		const parsed = validated.data;
		if (!context.graph.get(parsed.id)) continue;
		const pkgChangelogs = [];
		for (const id of parsed.changelogIds ?? []) {
			const entry = changelogs.get(id);
			if (entry) pkgChangelogs.push(entry);
		}
		packages.set(parsed.id, {
			changelogs: pkgChangelogs,
			updated: parsed.updated
		});
	}
	const plan = {
		options,
		changelogs,
		packages
	};
	for (const plugin of context.plugins) await handlePluginError(plugin, "initPublishPlan", () => plugin.initPublishPlan?.call(context, {
		lock,
		plan
	}));
	return plan;
}
function resolvePublishTargets(plan) {
	/** the iteration order = publish order */
	const orderedMap = /* @__PURE__ */ new Map();
	let lastSplitIndex = -1;
	/** package id -> true while scanning hard wait, false while scanning optional wait */
	const stack = /* @__PURE__ */ new Map();
	function scan(id) {
		const preflight = plan.packages.get(id)?.preflight;
		if (!preflight || !preflight.shouldPublish) return;
		switch (stack.get(id)) {
			case true: throw new Error(`circular reference of deps: ${[...stack.keys(), id].join(" -> ")}`);
			case false: return;
		}
		let ordered = orderedMap.get(id);
		if (ordered) return ordered;
		let split = false;
		if (preflight.wait) {
			stack.set(id, true);
			for (const dep of preflight.wait) {
				const ordered = scan(dep);
				split ||= ordered !== void 0 && ordered.index >= lastSplitIndex;
			}
		}
		if (preflight.optionalWait) {
			stack.set(id, false);
			for (const dep of preflight.optionalWait) {
				const ordered = scan(dep);
				split ||= ordered !== void 0 && ordered.index >= lastSplitIndex;
			}
		}
		stack.delete(id);
		ordered = {
			split,
			index: orderedMap.size
		};
		if (split) lastSplitIndex = ordered.index;
		orderedMap.set(id, ordered);
		return ordered;
	}
	for (const id of plan.packages.keys()) scan(id);
	return orderedMap;
}
async function runPublishPlan(context, plan) {
	const { dryRun = false, unstable_maxChunk = 5 } = plan.options;
	for (const plugin of context.plugins) await handlePluginError(plugin, "beforePublishAll", () => plugin.beforePublishAll?.call(context, { plan }));
	async function publish(pkg) {
		if (dryRun) return { type: "published" };
		try {
			for (const plugin of context.plugins) if (await handlePluginError(plugin, "willPublish", () => plugin.willPublish?.call(context, { pkg })) === false) return { type: "skipped" };
			for (const plugin of context.plugins) {
				const publishResult = await handlePluginError(plugin, "publish", () => plugin.publish?.call(context, {
					pkg,
					plan
				}));
				if (publishResult) return publishResult;
			}
		} catch (e) {
			return {
				type: "failed",
				error: e instanceof Error ? e.message : String(e)
			};
		}
		return {
			type: "failed",
			error: `There is no plugin to publish package "${pkg.id}", please make sure the package has a supported provider plugin.`
		};
	}
	let promises = [];
	for (const [id, { split }] of resolvePublishTargets(plan)) {
		const pkg = context.graph.get(id);
		const packagePlan = plan.packages.get(pkg.id);
		if (!pkg || !packagePlan) continue;
		if (split || promises.length >= unstable_maxChunk) {
			await Promise.all(promises);
			promises = [];
		}
		promises.push(publish(pkg).then(async (result) => {
			packagePlan.publishResult = result;
			if (result.type === "skipped") return;
			for (const plugin of context.plugins) await handlePluginError(plugin, "afterPublish", () => plugin.afterPublish?.call(context, {
				pkg,
				plan
			}));
		}));
	}
	await Promise.all(promises);
	for (const packagePlan of plan.packages.values()) packagePlan.publishResult ??= { type: "skipped" };
	for (const plugin of context.plugins) await handlePluginError(plugin, "afterPublishAll", () => plugin.afterPublishAll?.call(context, { plan }));
}
async function runPreflights(context, plan) {
	const { graph } = context;
	await Promise.all(Array.from(plan.packages, async ([id, packagePlan]) => {
		const pkg = graph.get(id);
		packagePlan.preflight = { shouldPublish: false };
		if (!packagePlan.updated) return;
		for (const plugin of context.plugins) {
			const res = await handlePluginError(plugin, "publishPreflight", () => plugin.publishPreflight?.call(context, {
				pkg,
				plan
			}));
			if (res) {
				packagePlan.preflight = res;
				break;
			}
		}
	}));
	if (plan.options.packages) {
		const only = /* @__PURE__ */ new Set();
		function addOnly(id) {
			if (only.has(id)) return;
			const preflight = plan.packages.get(id)?.preflight;
			if (!preflight) return;
			only.add(id);
			if (preflight.wait) for (const dep of preflight.wait) addOnly(dep);
			if (preflight.optionalWait) for (const dep of preflight.optionalWait) addOnly(dep);
		}
		for (const name of plan.options.packages) for (const pkg of graph.getByName(name)) addOnly(pkg.id);
		for (const [id, packagePlan] of plan.packages) {
			if (only.has(id)) continue;
			packagePlan.preflight.shouldPublish = false;
		}
	}
	for (const plugin of context.plugins) await handlePluginError(plugin, "afterPreflight", () => plugin.afterPreflight?.call(context, { plan }));
}
async function publishPlanStatus(plan, context) {
	for (const plugin of context.plugins) {
		const status = await handlePluginError(plugin, "resolvePlanStatus", () => plugin.resolvePlanStatus?.call(context, { plan }));
		if (Array.isArray(status) && await somePromise(status, (v) => v === "pending") || status === "pending") return {
			status: "pending",
			reason: `Plugin "${plugin.name}" has pending tasks`
		};
	}
	return { status: "success" };
}
//#endregion
export { createDraft as a, parseChangelogFile as c, runPublishPlan as i, readChangelogEntries as l, publishPlanStatus as n, PublishLock as o, runPreflights as r, parsePublishLock as s, initPublishPlan as t };
