// tegami/dist/semver-DrtaCCZK.js
import { diff, inc, parse } from "semver";
//#region src/utils/semver.ts
function formatNpmDistTag(distTag) {
	return distTag && distTag !== "latest" ? ` (${distTag})` : "";
}
function formatPackageVersion(name, version, distTag) {
	let out = name;
	if (version) out += `@${version}`;
	out += formatNpmDistTag(distTag);
	return out;
}
const WEIGHTS = {
	major: 3,
	premajor: 3,
	minor: 2,
	preminor: 2,
	patch: 1,
	prepatch: 1
};
const DEPTH = {
	major: 1,
	minor: 2,
	patch: 3
};
const PRE = {
	major: "premajor",
	minor: "preminor",
	patch: "prepatch"
};
function diffWeight(from, to) {
	const d = diff(from, to);
	if (!d) return 0;
	switch (d) {
		case "release":
		case "prerelease": return 4;
		default: return WEIGHTS[d];
	}
}
function maxBump(a, b) {
	if (WEIGHTS[a] > WEIGHTS[b]) return a;
	return b;
}
function bumpDepth(type) {
	return DEPTH[type];
}
function bumpVersion(version, type, prerelease) {
	let next = version;
	const parsed = parse(version);
	if (!parsed) next = null;
	else if (prerelease) {
		if (parsed.prerelease[0] === prerelease) next = type ? inc(parsed, "prerelease", prerelease) : version;
		else if (parsed.prerelease[0]) next = inc(parsed, "prerelease", prerelease);
		else if (type) next = inc(parsed, PRE[type], prerelease);
	} else if (type) next = inc(parsed, type);
	else if (parsed.prerelease.length > 0) next = inc(parsed, "release");
	if (!next) throw new Error(`Invalid semver version: ${version}`);
	return next;
}
//#endregion
export { formatPackageVersion as a, formatNpmDistTag as i, bumpVersion as n, maxBump as o, diffWeight as r, bumpDepth as t };
