// tegami/dist/shared-BraxbVVG.js
import { t as _accessExpressionAsString } from "./_accessExpressionAsString-DAz6bHI_.js";
import { n as _validateReport, t as _createStandardSchema } from "./_createStandardSchema-DFfVlh8x.js";
import { stringify } from "yaml";
//#region src/changelog/shared.ts
const validateChangelogFrontmatter = (() => {
	const _io0 = (input) => (void 0 === input.subject || "string" === typeof input.subject) && null !== input.packages && (void 0 === input.packages || Array.isArray(input.packages) && input.packages.every((elem) => "string" === typeof elem) || "object" === typeof input.packages && null !== input.packages && false === Array.isArray(input.packages) && _io1(input.packages));
	const _io1 = (input) => Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return void 0 !== value && (null === value || "major" === value || "minor" === value || "patch" === value || "object" === typeof value && null !== value && false === Array.isArray(value) && _io2(value));
	});
	const _io2 = (input) => (void 0 === input.type || "major" === input.type || "minor" === input.type || "patch" === input.type) && (void 0 === input.replay || Array.isArray(input.replay) && input.replay.every((elem) => "string" === typeof elem && 1 <= elem.length));
	const _vo0 = (input, _path, _exceptionable = true) => [void 0 === input.subject || "string" === typeof input.subject || _report(_exceptionable, {
		path: _path + ".subject",
		expected: "(string | undefined)",
		value: input.subject
	}), (null !== input.packages || _report(_exceptionable, {
		path: _path + ".packages",
		expected: "(Array<string> | Record<string, BumpType | ChangelogPackageConfig | null> | undefined)",
		value: input.packages
	})) && (void 0 === input.packages || Array.isArray(input.packages) && input.packages.map((elem, _index3) => "string" === typeof elem || _report(_exceptionable, {
		path: _path + ".packages[" + _index3 + "]",
		expected: "string",
		value: elem
	})).every((flag) => flag) || "object" === typeof input.packages && null !== input.packages && false === Array.isArray(input.packages) && _vo1(input.packages, _path + ".packages", _exceptionable) || _report(_exceptionable, {
		path: _path + ".packages",
		expected: "(Array<string> | Record<string, BumpType | ChangelogPackageConfig | null> | undefined)",
		value: input.packages
	}) || _report(_exceptionable, {
		path: _path + ".packages",
		expected: "(Array<string> | Record<string, BumpType | ChangelogPackageConfig | null> | undefined)",
		value: input.packages
	}))].every((flag) => flag);
	const _vo1 = (input, _path, _exceptionable = true) => [false === _exceptionable || Object.keys(input).map((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return (void 0 !== value || _report(_exceptionable, {
			path: _path + _accessExpressionAsString(key),
			expected: "(\"major\" | \"minor\" | \"patch\" | ChangelogPackageConfig | null)",
			value
		})) && (null === value || "major" === value || "minor" === value || "patch" === value || ("object" === typeof value && null !== value && false === Array.isArray(value) || _report(_exceptionable, {
			path: _path + _accessExpressionAsString(key),
			expected: "(\"major\" | \"minor\" | \"patch\" | ChangelogPackageConfig | null)",
			value
		})) && _vo2(value, _path + _accessExpressionAsString(key), _exceptionable) || _report(_exceptionable, {
			path: _path + _accessExpressionAsString(key),
			expected: "(\"major\" | \"minor\" | \"patch\" | ChangelogPackageConfig | null)",
			value
		}));
	}).every((flag) => flag)].every((flag) => flag);
	const _vo2 = (input, _path, _exceptionable = true) => [void 0 === input.type || "major" === input.type || "minor" === input.type || "patch" === input.type || _report(_exceptionable, {
		path: _path + ".type",
		expected: "(\"major\" | \"minor\" | \"patch\" | undefined)",
		value: input.type
	}), void 0 === input.replay || (Array.isArray(input.replay) || _report(_exceptionable, {
		path: _path + ".replay",
		expected: "(Array<string & MinLength<1>> | undefined)",
		value: input.replay
	})) && input.replay.map((elem, _index4) => "string" === typeof elem && (1 <= elem.length || _report(_exceptionable, {
		path: _path + ".replay[" + _index4 + "]",
		expected: "string & MinLength<1>",
		value: elem
	})) || _report(_exceptionable, {
		path: _path + ".replay[" + _index4 + "]",
		expected: "(string & MinLength<1>)",
		value: elem
	})).every((flag) => flag) || _report(_exceptionable, {
		path: _path + ".replay",
		expected: "(Array<string & MinLength<1>> | undefined)",
		value: input.replay
	})].every((flag) => flag);
	const __is = (input) => "object" === typeof input && null !== input && false === Array.isArray(input) && _io0(input);
	let errors;
	let _report;
	return _createStandardSchema((input) => {
		if (false === __is(input)) {
			errors = [];
			_report = _validateReport(errors);
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input && false === Array.isArray(input) || _report(true, {
				path: _path + "",
				expected: "ChangelogFrontmatter",
				value: input
			})) && _vo0(input, _path + "", true) || _report(true, {
				path: _path + "",
				expected: "ChangelogFrontmatter",
				value: input
			}))(input, "$input", true);
			const success = 0 === errors.length;
			return success ? {
				success,
				data: input
			} : {
				success,
				errors,
				data: input
			};
		}
		return {
			success: true,
			data: input
		};
	});
})();
function renderChangelog(frontmatter, body) {
	return [
		"---",
		stringify(frontmatter, { lineWidth: 0 }).trim(),
		"---",
		"",
		body.trim(),
		""
	].join("\n");
}
function getPackageBumps(graph, entry) {
	const packageBumps = /* @__PURE__ */ new Map();
	for (const [name, config] of entry.packages) {
		if (!config.type) continue;
		for (const pkg of graph.getByName(name)) packageBumps.set(pkg, config.type);
	}
	return packageBumps;
}
//#endregion
export { renderChangelog as n, validateChangelogFrontmatter as r, getPackageBumps as t };
