// tegami/dist/plugins/cargo.js
import { n as execFailure, r as fetchFailure } from "../error-CAsrGb6e.js";
import { t as _accessExpressionAsString } from "../_accessExpressionAsString-DAz6bHI_.js";
import { t as _assertGuard } from "../_assertGuard-cQ6dIdMj.js";
import { n as WorkspacePackage } from "../graph-DaJ28Y4f.js";
import fs from "node:fs/promises";
import path from "node:path";
import { x } from "tinyexec";
import * as semver$1 from "semver";
import { glob } from "tinyglobby";
import initToml, { edit, parse as parse$1 } from "@rainbowatcher/toml-edit-js";
//#region src/plugins/cargo/schema.ts
const assertCargoManifest = (() => {
	const _io0 = (input) => (void 0 === input["package"] || "object" === typeof input["package"] && null !== input["package"] && _io1(input["package"])) && (void 0 === input.workspace || "object" === typeof input.workspace && null !== input.workspace && false === Array.isArray(input.workspace) && _io3(input.workspace)) && (void 0 === input.dependencies || "object" === typeof input.dependencies && null !== input.dependencies && false === Array.isArray(input.dependencies) && _io5(input.dependencies)) && (void 0 === input["dev-dependencies"] || "object" === typeof input["dev-dependencies"] && null !== input["dev-dependencies"] && false === Array.isArray(input["dev-dependencies"]) && _io5(input["dev-dependencies"])) && (void 0 === input["build-dependencies"] || "object" === typeof input["build-dependencies"] && null !== input["build-dependencies"] && false === Array.isArray(input["build-dependencies"]) && _io5(input["build-dependencies"])) && (void 0 === input.target || "object" === typeof input.target && null !== input.target && false === Array.isArray(input.target) && _io10(input.target));
	const _io1 = (input) => "string" === typeof input.name && null !== input.version && void 0 !== input.version && ("string" === typeof input.version || "object" === typeof input.version && null !== input.version && _io2(input.version)) && null !== input.publish && (void 0 === input.publish || "boolean" === typeof input.publish || "object" === typeof input.publish && null !== input.publish && _io2(input.publish));
	const _io2 = (input) => true === input.workspace;
	const _io3 = (input) => (void 0 === input.members || Array.isArray(input.members) && input.members.every((elem) => "string" === typeof elem)) && (void 0 === input.exclude || Array.isArray(input.exclude) && input.exclude.every((elem) => "string" === typeof elem)) && (void 0 === input["package"] || "object" === typeof input["package"] && null !== input["package"] && false === Array.isArray(input["package"]) && _io4(input["package"])) && (void 0 === input.dependencies || "object" === typeof input.dependencies && null !== input.dependencies && false === Array.isArray(input.dependencies) && _io5(input.dependencies)) && (void 0 === input["dev-dependencies"] || "object" === typeof input["dev-dependencies"] && null !== input["dev-dependencies"] && false === Array.isArray(input["dev-dependencies"]) && _io5(input["dev-dependencies"])) && (void 0 === input["build-dependencies"] || "object" === typeof input["build-dependencies"] && null !== input["build-dependencies"] && false === Array.isArray(input["build-dependencies"]) && _io5(input["build-dependencies"]));
	const _io4 = (input) => (void 0 === input.version || "string" === typeof input.version) && (void 0 === input.publish || "boolean" === typeof input.publish);
	const _io5 = (input) => Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return null !== value && void 0 !== value && ("string" === typeof value || "object" === typeof value && null !== value && _iu0(value));
	});
	const _io6 = (input) => true === input.workspace && (void 0 === input["package"] || "string" === typeof input["package"]) && (void 0 === input.features || Array.isArray(input.features) && input.features.every((elem) => "string" === typeof elem)) && (void 0 === input.optional || "boolean" === typeof input.optional) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"]);
	const _io7 = (input) => "string" === typeof input.path && (void 0 === input.version || "string" === typeof input.version) && (void 0 === input["package"] || "string" === typeof input["package"]) && (void 0 === input.features || Array.isArray(input.features) && input.features.every((elem) => "string" === typeof elem)) && (void 0 === input.optional || "boolean" === typeof input.optional) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"]);
	const _io8 = (input) => "string" === typeof input.git && (void 0 === input.branch || "string" === typeof input.branch) && (void 0 === input.tag || "string" === typeof input.tag) && (void 0 === input.rev || "string" === typeof input.rev) && (void 0 === input.version || "string" === typeof input.version) && (void 0 === input["package"] || "string" === typeof input["package"]) && (void 0 === input.features || Array.isArray(input.features) && input.features.every((elem) => "string" === typeof elem)) && (void 0 === input.optional || "boolean" === typeof input.optional) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"]);
	const _io9 = (input) => "string" === typeof input.version && (void 0 === input.registry || "string" === typeof input.registry) && (void 0 === input["package"] || "string" === typeof input["package"]) && (void 0 === input.features || Array.isArray(input.features) && input.features.every((elem) => "string" === typeof elem)) && (void 0 === input.optional || "boolean" === typeof input.optional) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"]);
	const _io10 = (input) => Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return "object" === typeof value && null !== value && false === Array.isArray(value) && _io11(value);
	});
	const _io11 = (input) => (void 0 === input.dependencies || "object" === typeof input.dependencies && null !== input.dependencies && false === Array.isArray(input.dependencies) && _io5(input.dependencies)) && (void 0 === input["dev-dependencies"] || "object" === typeof input["dev-dependencies"] && null !== input["dev-dependencies"] && false === Array.isArray(input["dev-dependencies"]) && _io5(input["dev-dependencies"])) && (void 0 === input["build-dependencies"] || "object" === typeof input["build-dependencies"] && null !== input["build-dependencies"] && false === Array.isArray(input["build-dependencies"]) && _io5(input["build-dependencies"]));
	const _iu0 = (input) => (() => {
		if (void 0 !== input.workspace) return _io6(input);
		else if (void 0 !== input.path) return _io7(input);
		else if (void 0 !== input.git) return _io8(input);
		else return _io9(input);
	})();
	const _ao0 = (input, _path, _exceptionable = true) => (void 0 === input["package"] || ("object" === typeof input["package"] && null !== input["package"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(__type | undefined)",
		value: input["package"]
	}, _errorFactory)) && _ao1(input["package"], _path + "[\"package\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(__type | undefined)",
		value: input["package"]
	}, _errorFactory)) && (void 0 === input.workspace || ("object" === typeof input.workspace && null !== input.workspace && false === Array.isArray(input.workspace) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".workspace",
		expected: "(CargoWorkspace | undefined)",
		value: input.workspace
	}, _errorFactory)) && _ao3(input.workspace, _path + ".workspace", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".workspace",
		expected: "(CargoWorkspace | undefined)",
		value: input.workspace
	}, _errorFactory)) && (void 0 === input.dependencies || ("object" === typeof input.dependencies && null !== input.dependencies && false === Array.isArray(input.dependencies) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".dependencies",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input.dependencies
	}, _errorFactory)) && _ao5(input.dependencies, _path + ".dependencies", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".dependencies",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input.dependencies
	}, _errorFactory)) && (void 0 === input["dev-dependencies"] || ("object" === typeof input["dev-dependencies"] && null !== input["dev-dependencies"] && false === Array.isArray(input["dev-dependencies"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"dev-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["dev-dependencies"]
	}, _errorFactory)) && _ao5(input["dev-dependencies"], _path + "[\"dev-dependencies\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"dev-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["dev-dependencies"]
	}, _errorFactory)) && (void 0 === input["build-dependencies"] || ("object" === typeof input["build-dependencies"] && null !== input["build-dependencies"] && false === Array.isArray(input["build-dependencies"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"build-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["build-dependencies"]
	}, _errorFactory)) && _ao5(input["build-dependencies"], _path + "[\"build-dependencies\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"build-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["build-dependencies"]
	}, _errorFactory)) && (void 0 === input.target || ("object" === typeof input.target && null !== input.target && false === Array.isArray(input.target) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".target",
		expected: "(Record<string, CargoTargetSection> | undefined)",
		value: input.target
	}, _errorFactory)) && _ao10(input.target, _path + ".target", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".target",
		expected: "(Record<string, CargoTargetSection> | undefined)",
		value: input.target
	}, _errorFactory));
	const _ao1 = (input, _path, _exceptionable = true) => ("string" === typeof input.name || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".name",
		expected: "string",
		value: input.name
	}, _errorFactory)) && (null !== input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(CargoInherit | string)",
		value: input.version
	}, _errorFactory)) && (void 0 !== input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(CargoInherit | string)",
		value: input.version
	}, _errorFactory)) && ("string" === typeof input.version || ("object" === typeof input.version && null !== input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(CargoInherit | string)",
		value: input.version
	}, _errorFactory)) && _ao2(input.version, _path + ".version", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(CargoInherit | string)",
		value: input.version
	}, _errorFactory)) && (null !== input.publish || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".publish",
		expected: "(CargoInherit | boolean | undefined)",
		value: input.publish
	}, _errorFactory)) && (void 0 === input.publish || "boolean" === typeof input.publish || ("object" === typeof input.publish && null !== input.publish || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".publish",
		expected: "(CargoInherit | boolean | undefined)",
		value: input.publish
	}, _errorFactory)) && _ao2(input.publish, _path + ".publish", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".publish",
		expected: "(CargoInherit | boolean | undefined)",
		value: input.publish
	}, _errorFactory));
	const _ao2 = (input, _path, _exceptionable = true) => true === input.workspace || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".workspace",
		expected: "true",
		value: input.workspace
	}, _errorFactory);
	const _ao3 = (input, _path, _exceptionable = true) => (void 0 === input.members || (Array.isArray(input.members) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".members",
		expected: "(Array<string> | undefined)",
		value: input.members
	}, _errorFactory)) && input.members.every((elem, _index7) => "string" === typeof elem || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".members[" + _index7 + "]",
		expected: "string",
		value: elem
	}, _errorFactory)) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".members",
		expected: "(Array<string> | undefined)",
		value: input.members
	}, _errorFactory)) && (void 0 === input.exclude || (Array.isArray(input.exclude) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".exclude",
		expected: "(Array<string> | undefined)",
		value: input.exclude
	}, _errorFactory)) && input.exclude.every((elem, _index8) => "string" === typeof elem || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".exclude[" + _index8 + "]",
		expected: "string",
		value: elem
	}, _errorFactory)) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".exclude",
		expected: "(Array<string> | undefined)",
		value: input.exclude
	}, _errorFactory)) && (void 0 === input["package"] || ("object" === typeof input["package"] && null !== input["package"] && false === Array.isArray(input["package"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(__type.o1 | undefined)",
		value: input["package"]
	}, _errorFactory)) && _ao4(input["package"], _path + "[\"package\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(__type.o1 | undefined)",
		value: input["package"]
	}, _errorFactory)) && (void 0 === input.dependencies || ("object" === typeof input.dependencies && null !== input.dependencies && false === Array.isArray(input.dependencies) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".dependencies",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input.dependencies
	}, _errorFactory)) && _ao5(input.dependencies, _path + ".dependencies", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".dependencies",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input.dependencies
	}, _errorFactory)) && (void 0 === input["dev-dependencies"] || ("object" === typeof input["dev-dependencies"] && null !== input["dev-dependencies"] && false === Array.isArray(input["dev-dependencies"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"dev-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["dev-dependencies"]
	}, _errorFactory)) && _ao5(input["dev-dependencies"], _path + "[\"dev-dependencies\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"dev-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["dev-dependencies"]
	}, _errorFactory)) && (void 0 === input["build-dependencies"] || ("object" === typeof input["build-dependencies"] && null !== input["build-dependencies"] && false === Array.isArray(input["build-dependencies"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"build-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["build-dependencies"]
	}, _errorFactory)) && _ao5(input["build-dependencies"], _path + "[\"build-dependencies\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"build-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["build-dependencies"]
	}, _errorFactory));
	const _ao4 = (input, _path, _exceptionable = true) => (void 0 === input.version || "string" === typeof input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(string | undefined)",
		value: input.version
	}, _errorFactory)) && (void 0 === input.publish || "boolean" === typeof input.publish || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".publish",
		expected: "(boolean | undefined)",
		value: input.publish
	}, _errorFactory));
	const _ao5 = (input, _path, _exceptionable = true) => false === _exceptionable || Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return (null !== value || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "(CargoGitDependency | CargoPathDependency | CargoRegistryDependency | CargoWorkspaceDependency | string)",
			value
		}, _errorFactory)) && (void 0 !== value || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "(CargoGitDependency | CargoPathDependency | CargoRegistryDependency | CargoWorkspaceDependency | string)",
			value
		}, _errorFactory)) && ("string" === typeof value || ("object" === typeof value && null !== value || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "(CargoGitDependency | CargoPathDependency | CargoRegistryDependency | CargoWorkspaceDependency | string)",
			value
		}, _errorFactory)) && _au0(value, _path + _accessExpressionAsString(key), _exceptionable) || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "(CargoGitDependency | CargoPathDependency | CargoRegistryDependency | CargoWorkspaceDependency | string)",
			value
		}, _errorFactory));
	});
	const _ao6 = (input, _path, _exceptionable = true) => (true === input.workspace || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".workspace",
		expected: "true",
		value: input.workspace
	}, _errorFactory)) && (void 0 === input["package"] || "string" === typeof input["package"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(string | undefined)",
		value: input["package"]
	}, _errorFactory)) && (void 0 === input.features || (Array.isArray(input.features) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && input.features.every((elem, _index9) => "string" === typeof elem || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features[" + _index9 + "]",
		expected: "string",
		value: elem
	}, _errorFactory)) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && (void 0 === input.optional || "boolean" === typeof input.optional || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".optional",
		expected: "(boolean | undefined)",
		value: input.optional
	}, _errorFactory)) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"default-features\"]",
		expected: "(boolean | undefined)",
		value: input["default-features"]
	}, _errorFactory));
	const _ao7 = (input, _path, _exceptionable = true) => ("string" === typeof input.path || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".path",
		expected: "string",
		value: input.path
	}, _errorFactory)) && (void 0 === input.version || "string" === typeof input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(string | undefined)",
		value: input.version
	}, _errorFactory)) && (void 0 === input["package"] || "string" === typeof input["package"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(string | undefined)",
		value: input["package"]
	}, _errorFactory)) && (void 0 === input.features || (Array.isArray(input.features) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && input.features.every((elem, _index10) => "string" === typeof elem || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features[" + _index10 + "]",
		expected: "string",
		value: elem
	}, _errorFactory)) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && (void 0 === input.optional || "boolean" === typeof input.optional || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".optional",
		expected: "(boolean | undefined)",
		value: input.optional
	}, _errorFactory)) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"default-features\"]",
		expected: "(boolean | undefined)",
		value: input["default-features"]
	}, _errorFactory));
	const _ao8 = (input, _path, _exceptionable = true) => ("string" === typeof input.git || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".git",
		expected: "string",
		value: input.git
	}, _errorFactory)) && (void 0 === input.branch || "string" === typeof input.branch || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".branch",
		expected: "(string | undefined)",
		value: input.branch
	}, _errorFactory)) && (void 0 === input.tag || "string" === typeof input.tag || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".tag",
		expected: "(string | undefined)",
		value: input.tag
	}, _errorFactory)) && (void 0 === input.rev || "string" === typeof input.rev || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".rev",
		expected: "(string | undefined)",
		value: input.rev
	}, _errorFactory)) && (void 0 === input.version || "string" === typeof input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "(string | undefined)",
		value: input.version
	}, _errorFactory)) && (void 0 === input["package"] || "string" === typeof input["package"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(string | undefined)",
		value: input["package"]
	}, _errorFactory)) && (void 0 === input.features || (Array.isArray(input.features) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && input.features.every((elem, _index11) => "string" === typeof elem || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features[" + _index11 + "]",
		expected: "string",
		value: elem
	}, _errorFactory)) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && (void 0 === input.optional || "boolean" === typeof input.optional || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".optional",
		expected: "(boolean | undefined)",
		value: input.optional
	}, _errorFactory)) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"default-features\"]",
		expected: "(boolean | undefined)",
		value: input["default-features"]
	}, _errorFactory));
	const _ao9 = (input, _path, _exceptionable = true) => ("string" === typeof input.version || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".version",
		expected: "string",
		value: input.version
	}, _errorFactory)) && (void 0 === input.registry || "string" === typeof input.registry || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".registry",
		expected: "(string | undefined)",
		value: input.registry
	}, _errorFactory)) && (void 0 === input["package"] || "string" === typeof input["package"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"package\"]",
		expected: "(string | undefined)",
		value: input["package"]
	}, _errorFactory)) && (void 0 === input.features || (Array.isArray(input.features) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && input.features.every((elem, _index12) => "string" === typeof elem || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features[" + _index12 + "]",
		expected: "string",
		value: elem
	}, _errorFactory)) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".features",
		expected: "(Array<string> | undefined)",
		value: input.features
	}, _errorFactory)) && (void 0 === input.optional || "boolean" === typeof input.optional || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".optional",
		expected: "(boolean | undefined)",
		value: input.optional
	}, _errorFactory)) && (void 0 === input["default-features"] || "boolean" === typeof input["default-features"] || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"default-features\"]",
		expected: "(boolean | undefined)",
		value: input["default-features"]
	}, _errorFactory));
	const _ao10 = (input, _path, _exceptionable = true) => false === _exceptionable || Object.keys(input).every((key) => {
		const value = input[key];
		if (void 0 === value) return true;
		return ("object" === typeof value && null !== value && false === Array.isArray(value) || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "CargoTargetSection",
			value
		}, _errorFactory)) && _ao11(value, _path + _accessExpressionAsString(key), _exceptionable) || _assertGuard(_exceptionable, {
			method: "typia.createAssert",
			path: _path + _accessExpressionAsString(key),
			expected: "CargoTargetSection",
			value
		}, _errorFactory);
	});
	const _ao11 = (input, _path, _exceptionable = true) => (void 0 === input.dependencies || ("object" === typeof input.dependencies && null !== input.dependencies && false === Array.isArray(input.dependencies) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".dependencies",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input.dependencies
	}, _errorFactory)) && _ao5(input.dependencies, _path + ".dependencies", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + ".dependencies",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input.dependencies
	}, _errorFactory)) && (void 0 === input["dev-dependencies"] || ("object" === typeof input["dev-dependencies"] && null !== input["dev-dependencies"] && false === Array.isArray(input["dev-dependencies"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"dev-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["dev-dependencies"]
	}, _errorFactory)) && _ao5(input["dev-dependencies"], _path + "[\"dev-dependencies\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"dev-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["dev-dependencies"]
	}, _errorFactory)) && (void 0 === input["build-dependencies"] || ("object" === typeof input["build-dependencies"] && null !== input["build-dependencies"] && false === Array.isArray(input["build-dependencies"]) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"build-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["build-dependencies"]
	}, _errorFactory)) && _ao5(input["build-dependencies"], _path + "[\"build-dependencies\"]", _exceptionable) || _assertGuard(_exceptionable, {
		method: "typia.createAssert",
		path: _path + "[\"build-dependencies\"]",
		expected: "(Record<string, CargoDependency> | undefined)",
		value: input["build-dependencies"]
	}, _errorFactory));
	const _au0 = (input, _path, _exceptionable = true) => (() => {
		if (void 0 !== input.workspace) return _ao6(input, _path, _exceptionable);
		else if (void 0 !== input.path) return _ao7(input, _path, _exceptionable);
		else if (void 0 !== input.git) return _ao8(input, _path, _exceptionable);
		else return _ao9(input, _path, _exceptionable);
	})();
	const __is = (input) => "object" === typeof input && null !== input && false === Array.isArray(input) && _io0(input);
	let _errorFactory;
	return (input, errorFactory) => {
		if (false === __is(input)) {
			_errorFactory = errorFactory;
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input && false === Array.isArray(input) || _assertGuard(true, {
				method: "typia.createAssert",
				path: _path + "",
				expected: "CargoManifest",
				value: input
			}, _errorFactory)) && _ao0(input, _path + "", true) || _assertGuard(true, {
				method: "typia.createAssert",
				path: _path + "",
				expected: "CargoManifest",
				value: input
			}, _errorFactory))(input, "$input", true);
		}
		return input;
	};
})();
//#endregion
//#region src/plugins/cargo.ts
const DEP_FIELDS = [
	"dependencies",
	"dev-dependencies",
	"build-dependencies"
];
var CargoToml = class {
	path;
	content;
	data;
	dependencies;
	workspace;
	originalContent;
	constructor(path, content, data) {
		this.path = path;
		this.content = content;
		this.data = data;
		this.originalContent = content;
	}
	listDependencies(graph) {
		if (this.dependencies) return this.dependencies;
		const out = this.dependencies = [];
		const scan = (obj, prefix = []) => {
			for (const field of DEP_FIELDS) {
				const table = obj[field];
				if (!table) continue;
				const tablePath = [...prefix, field];
				for (const [key, spec] of Object.entries(table)) {
					const { linked, range, setRange } = resolveLinkedDep(this, graph, tablePath, key, spec) ?? {};
					out.push({
						path: [...tablePath, key],
						spec,
						resolved: linked,
						range,
						setRange: setRange && ((v) => {
							table[key] = setRange(v);
						})
					});
				}
			}
		};
		scan(this.data);
		if (this.data.target) for (const [key, config] of Object.entries(this.data.target)) scan(config, ["target", key]);
		if (this.data.workspace) scan(this.data.workspace);
		return out;
	}
	patch(path, value) {
		this.content = edit(this.content, path, value);
	}
	async write() {
		if (this.content === this.originalContent) return;
		await fs.writeFile(this.path, this.content);
	}
};
var CargoPackage = class extends WorkspacePackage {
	path;
	file;
	manager = "cargo";
	manifest;
	constructor(path, file) {
		super();
		this.path = path;
		this.file = file;
		this.manifest = file.data;
	}
	get name() {
		return this.manifest.package.name;
	}
	get version() {
		const packageVersion = this.manifest.package.version;
		if (typeof packageVersion === "string") return packageVersion;
		const inherited = this.file.workspace?.data.workspace.package;
		if (packageVersion.workspace && inherited?.version) return inherited.version;
		throw new Error(`Invalid Cargo.toml in "${this.path}".`);
	}
	setVersion(version) {
		const packageInfo = this.manifest.package;
		if (typeof packageInfo.version === "string") {
			packageInfo.version = version;
			this.file.patch("package.version", version);
			return;
		}
		if (packageInfo.version.workspace && this.file.workspace?.data.workspace.package) {
			this.file.workspace.data.workspace.package.version = version;
			this.file.workspace.patch("workspace.package.version", version);
			return;
		}
		throw new Error(`Invalid Cargo.toml in "${this.path}".`);
	}
};
function cargo({ updateLockFile = true, bumpDep: getBumpDepType } = {}) {
	return {
		name: "cargo",
		async init() {
			await initToml();
		},
		async resolve() {
			const graph = await resolveCargoGraph(this.cwd);
			if (graph.packages.size === 0) return;
			this.cargo = { graph };
			for (const pkg of graph.packages.values()) this.graph.add(pkg);
		},
		initDraft(plan) {
			if (!this.cargo) return;
			plan.addPolicy(depsPolicy(this, getBumpDepType));
		},
		async publishPreflight({ pkg }) {
			if (!(pkg instanceof CargoPackage) || !this.cargo) return;
			let shouldPublish = true;
			if (typeof pkg.manifest.package.publish === "boolean") shouldPublish = pkg.manifest.package.publish;
			else if (pkg.manifest.package.publish?.workspace) shouldPublish = pkg.file.workspace?.data.workspace?.package?.publish ?? shouldPublish;
			const wait = [];
			for (const { resolved } of pkg.file.listDependencies(this.cargo.graph)) if (resolved) wait.push(resolved.id);
			return {
				shouldPublish,
				wait
			};
		},
		resolvePlanStatus({ plan }) {
			if (!this.cargo) return;
			return Array.from(plan.packages, async ([id, { preflight }]) => {
				if (!preflight.shouldPublish) return;
				const pkg = this.graph.get(id);
				if (!(pkg instanceof CargoPackage)) return;
				if (!await isPackagePublished(pkg.name, pkg.version)) return "pending";
			});
		},
		async publish({ pkg }) {
			if (!(pkg instanceof CargoPackage)) return;
			const result = await x("cargo", ["publish"], { nodeOptions: { cwd: pkg.path } });
			if (result.exitCode !== 0) {
				if (/already exists|already published/i.test(`${result.stdout}\n${result.stderr}`)) return { type: "skipped" };
				return {
					type: "failed",
					error: execFailure(`Failed to publish ${pkg.name}@${pkg.version}.`, result).message
				};
			}
			return { type: "published" };
		},
		async applyDraft(draft) {
			if (!this.cargo) return;
			const graph = this.cargo.graph;
			for (const pkg of graph.packages.values()) {
				const bumped = draft.getPackageDraft(pkg.id)?.bumpVersion(pkg);
				if (bumped) pkg.setVersion(bumped);
			}
			for (const file of graph.files.values()) for (const { range, resolved, setRange } of file.listDependencies(graph)) {
				if (!resolved || !range || !setRange || semver$1.satisfies(resolved.version, range)) continue;
				let updatedRange;
				if (range.startsWith("^")) updatedRange = `^${resolved.version}`;
				else if (range.startsWith("~")) updatedRange = `~${resolved.version}`;
				else updatedRange = resolved.version;
				setRange(updatedRange);
			}
			await Promise.all(Array.from(graph.files.values(), (file) => file.write()));
		},
		async applyCliDraft() {
			if (!this.cargo || !updateLockFile) return;
			const result = await x("cargo", ["update", "--workspace"], { nodeOptions: { cwd: this.cwd } });
			if (result.exitCode !== 0) throw execFailure("Failed to update Cargo lock file", result);
		}
	};
}
function depsPolicy({ cargo }, getBumpDepType = ({ kind }) => {
	switch (kind) {
		case "dependencies": return "patch";
		case "build-dependencies":
		case "dev-dependencies": return false;
	}
}) {
	const cargoGraph = cargo.graph;
	const dependentMap = /* @__PURE__ */ new Map();
	for (const pkg of cargoGraph.packages.values()) for (const { resolved, path, spec, range } of pkg.file.listDependencies(cargoGraph)) {
		if (!resolved) continue;
		const refs = dependentMap.get(resolved.id);
		const kind = path[path.length - 2];
		const name = path[path.length - 1];
		if (refs) refs.push({
			dependent: pkg,
			kind,
			name,
			spec,
			version: range
		});
		else dependentMap.set(resolved.id, [{
			dependent: pkg,
			kind,
			name,
			spec,
			version: range
		}]);
	}
	return {
		id: "cargo:deps",
		onUpdate({ pkg, packageDraft: plan }) {
			if (!(pkg instanceof CargoPackage)) return;
			const deps = dependentMap.get(pkg.id);
			if (!deps) return;
			const bumped = plan.bumpVersion(pkg);
			if (!bumped) return;
			for (const dep of deps) {
				if (pkg.group?.options.syncBump && dep.dependent.group === pkg.group) continue;
				if (dep.version && semver$1.satisfies(bumped, dep.version)) continue;
				const bumpType = getBumpDepType(dep);
				if (bumpType === false) continue;
				this.bumpPackage(dep.dependent, {
					type: bumpType,
					reason: `update dependency "${pkg.name}"`
				});
			}
		}
	};
}
async function isPackagePublished(name, version) {
	const response = await fetch(`https://crates.io/api/v1/crates/${encodeURIComponent(name)}/${version}`);
	if (response.status === 200) return true;
	if (response.status === 404) return false;
	throw await fetchFailure(`Unable to validate ${name}@${version} against crates.io`, response);
}
async function buildEntry(dir) {
	try {
		const filePath = path.join(dir, "Cargo.toml");
		const content = await fs.readFile(filePath, "utf8");
		return new CargoToml(filePath, content, assertCargoManifest(parse$1(content)));
	} catch {
		return;
	}
}
async function resolveCargoGraph(cwd) {
	const out = {
		packages: /* @__PURE__ */ new Map(),
		files: /* @__PURE__ */ new Map()
	};
	const root = await buildEntry(cwd);
	if (!root) return out;
	out.files.set(root.path, root);
	if (root.data.package) {
		const pkg = new CargoPackage(cwd, root);
		out.packages.set(pkg.name, pkg);
	}
	if (root.data.workspace?.members) {
		root.workspace = root;
		const dirs = await expandWorkspaceMembers(cwd, root.data.workspace.members, root.data.workspace.exclude);
		await Promise.all(dirs.map(async (dir) => {
			const entry = await buildEntry(dir);
			if (!entry || !entry.data.package) return;
			entry.workspace = root;
			const pkg = new CargoPackage(dir, entry);
			out.files.set(entry.path, entry);
			out.packages.set(pkg.name, pkg);
		}));
	}
	return out;
}
async function expandWorkspaceMembers(cwd, members, exclude = []) {
	const patterns = members.filter((member) => member !== ".");
	if (patterns.length === 0) return [];
	return (await glob(patterns, {
		absolute: true,
		cwd,
		ignore: ["**/target/**", ...exclude],
		onlyDirectories: true,
		onlyFiles: false
	})).map((item) => {
		return item.endsWith(path.sep) ? item.slice(0, -1) : item;
	});
}
function resolveLinkedDep(file, graph, tablePath, key, spec) {
	if (typeof spec === "string") {
		const linked = graph.packages.get(key);
		if (!linked) return;
		return {
			linked,
			range: spec,
			setRange(v) {
				file.patch([...tablePath, key].join("."), v);
				return v;
			}
		};
	}
	if ("git" in spec) return;
	if ("workspace" in spec) {
		const kind = tablePath[tablePath.length - 1];
		const entry = file.workspace?.data.workspace?.[kind]?.[key];
		if (!entry) return;
		const resolved = resolveLinkedDep(file.workspace, graph, ["workspace", kind], key, entry);
		if (!resolved) return;
		delete resolved.setRange;
		return resolved;
	}
	if ("path" in spec) {
		const pkgPath = path.resolve(path.dirname(file.path), spec.path);
		for (const pkg of graph.packages.values()) {
			if (pkg.path !== pkgPath) continue;
			return {
				linked: pkg,
				range: spec.version,
				setRange(v) {
					file.patch([
						...tablePath,
						key,
						"version"
					].join("."), v);
					return {
						...spec,
						version: v
					};
				}
			};
		}
		return;
	}
	const linked = graph.packages.get(spec.package ?? key);
	if (!linked) return;
	return {
		linked,
		range: spec.version,
		setRange(v) {
			file.patch([
				...tablePath,
				key,
				"version"
			].join("."), v);
			return {
				...spec,
				version: v
			};
		}
	};
}
//#endregion
export { CargoPackage, cargo };
