// tegami/dist/plugins/github.js
import { t as changelogFilename } from "../generate-Cip5o2xy.js";
import { n as execFailure, o as cached, r as fetchFailure, s as isCI } from "../error-CAsrGb6e.js";
import { a as createDraft, l as readChangelogEntries } from "../publish-CQifflae.js";
import { git } from "./git.js";
import { i as versionRequestPlugin, n as formatPreview, r as resolveFileCommit, t as createAutoRelease } from "../version-request-CLEmlVnL.js";
import { readFile, writeFile } from "node:fs/promises";
import { basename, join, relative, resolve } from "node:path";
import { x } from "tinyexec";
import semver from "semver";
import { note, outro } from "@clack/prompts";
//#region src/plugins/github/api.ts
function parseGitHubRepo(repo) {
	const [owner, name] = repo.split("/", 2);
	if (!owner || !name) throw new Error(`Invalid GitHub repository: ${repo}`);
	return {
		owner,
		repo: name
	};
}
async function githubRequest(path, token, init = {}) {
	const headers = new Headers(init.headers);
	headers.set("Accept", "application/vnd.github+json");
	headers.set("X-GitHub-Api-Version", "2022-11-28");
	if (token) headers.set("Authorization", `Bearer ${token}`);
	return fetch(`https://api.github.com${path}`, {
		...init,
		headers
	});
}
async function releaseExistsByTag(repo, tag, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/releases/tags/${encodeURIComponent(tag)}`, token, { method: "HEAD" });
	if (response.status === 404) return false;
	if (!response.ok) throw await fetchFailure(`Failed to get GitHub release for ${tag}`, response);
	return true;
}
async function createRelease(options) {
	const { owner, repo } = parseGitHubRepo(options.repo);
	const response = await githubRequest(`/repos/${owner}/${repo}/releases`, options.token, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			tag_name: options.tag,
			name: options.title,
			body: options.notes,
			prerelease: options.prerelease ?? false
		})
	});
	if (!response.ok) throw await fetchFailure(`Failed to create GitHub release for ${options.tag}`, response);
}
async function findOpenPullRequest(repo, headBranch, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/pulls?${new URLSearchParams({
		head: `${owner}:${headBranch}`,
		state: "open"
	})}`, token);
	if (!response.ok) throw await fetchFailure("Failed to check for an existing version pull request", response);
	return (await response.json())[0]?.number;
}
async function updatePullRequest(repo, number, options) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/pulls/${number}`, options.token, {
		method: "PATCH",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			title: options.title,
			body: options.body
		})
	});
	if (!response.ok) throw await fetchFailure("Failed to update the version pull request", response);
}
async function createPullRequest(repo, options) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/pulls`, options.token, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			title: options.title,
			body: options.body,
			head: options.head,
			base: options.base
		})
	});
	if (!response.ok) throw await fetchFailure("Failed to create the version pull request", response);
}
async function listPullRequestsForCommit(repo, commitSha, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/commits/${commitSha}/pulls`, token);
	if (!response.ok) throw await fetchFailure(`Failed to list pull requests for commit ${commitSha.slice(0, 7)}`, response);
	return await response.json();
}
async function getPullRequest(repo, number, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/pulls/${number}`, token);
	if (!response.ok) throw await fetchFailure(`Failed to resolve pull request #${number}`, response);
	const data = await response.json();
	return {
		headRefName: data.head.ref,
		baseRefOid: data.base.sha,
		headRefOid: data.head.sha,
		headRepository: data.head.repo
	};
}
async function findIssueCommentByPrefix(repo, issueNumber, prefix, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	let page = 1;
	while (true) {
		const response = await githubRequest(`/repos/${owner}/${name}/issues/${issueNumber}/comments?per_page=100&page=${page}`, token);
		if (!response.ok) throw await fetchFailure("Failed to list pull request comments", response);
		const batch = await response.json();
		if (batch.length === 0) break;
		const comment = batch.find((comment) => comment.body.startsWith(prefix));
		if (comment) return comment.id;
		if (batch.length < 100) break;
		page += 1;
	}
}
async function updateIssueComment(repo, commentId, body, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/issues/comments/${commentId}`, token, {
		method: "PATCH",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ body })
	});
	if (!response.ok) throw await fetchFailure("Failed to update pull request comment", response);
}
async function createIssueComment(repo, issueNumber, body, token) {
	const { owner, repo: name } = parseGitHubRepo(repo);
	const response = await githubRequest(`/repos/${owner}/${name}/issues/${issueNumber}/comments`, token, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ body })
	});
	if (!response.ok) throw await fetchFailure("Failed to create pull request comment", response);
}
//#endregion
//#region src/plugins/github/cli.ts
const COMMENT_MARKER = "<!-- tegami -->";
function registerPrCli(cli) {
	cli.command("pr preview", { description: "show a pull request release preview and changelog guidance" }).option("artifact", {
		type: "string",
		description: "write preview markdown to a file"
	}).option("number", {
		type: "string",
		description: "pull request number"
	}).action(async ({ context, values }) => {
		const number = values.number ? parsePositiveInt(values.number, "--number") : void 0;
		const artifact = values.artifact;
		const body = await buildPrPreview(context, await createDraft(await readChangelogEntries(context), context), { number });
		if (artifact) {
			const artifactPath = resolve(context.cwd, artifact);
			await writeFile(artifactPath, body);
			if (!isCI()) {
				note(relative(context.cwd, artifactPath) || artifact, "Release preview");
				outro("Release preview ready.");
			}
			return;
		}
		if (isCI()) {
			process.stdout.write(`${body}\n`);
			return;
		}
		note(body, "Release preview");
		outro("Release preview ready.");
	});
	cli.command("pr comment", { description: "post the pull request release preview as a comment" }).positional("artifact").action(async ({ context, positionals }) => {
		await postPrComment(context, await readFile(positionals.artifact, "utf8"));
		outro("Pull request comment updated.");
	});
}
async function buildPrPreview(context, draft, options = {}) {
	const pullRequest = await resolvePullRequest(context, options);
	return formatPreview(context, draft, await listPullRequestChangelogFiles(context, pullRequest.baseSha, pullRequest.headSha), {
		"create-a-changelog-href": createChangelogUrl(context, pullRequest.headRepo, pullRequest.headRef, changelogFilename()),
		pr: "PR"
	});
}
async function postPrComment(context, body) {
	const { repo, token } = context.github ?? {};
	if (!repo) {
		outro("GitHub plugin context is required.");
		return;
	}
	const pr = await readPullRequestFromWorkflowRunEvent();
	if (!pr.found) {
		outro(pr.reason);
		return;
	}
	const markedBody = `${COMMENT_MARKER}\n${body}`;
	const existingId = await findIssueCommentByPrefix(repo, pr.number, COMMENT_MARKER, token);
	if (existingId) {
		await updateIssueComment(repo, existingId, markedBody, token);
		return;
	}
	await createIssueComment(repo, pr.number, markedBody, token);
}
async function resolvePullRequest(context, options) {
	const { repo, token } = context.github ?? {};
	if (!repo) throw new Error("GitHub plugin context is required.");
	if (options.number !== void 0) {
		const data = await getPullRequest(repo, options.number, token);
		return {
			headRepo: data.headRepository ? `${data.headRepository.owner.login}/${data.headRepository.name}` : repo,
			headRef: data.headRefName,
			baseSha: data.baseRefOid,
			headSha: data.headRefOid
		};
	}
	const eventPath = process.env.GITHUB_EVENT_PATH;
	if (eventPath) try {
		const pullRequest = JSON.parse(await readFile(eventPath, "utf8")).pull_request;
		if (pullRequest) return {
			headRepo: pullRequest.head.repo.full_name,
			headRef: pullRequest.head.ref,
			baseSha: pullRequest.base.sha,
			headSha: pullRequest.head.sha
		};
	} catch {}
	throw new Error("A pull request event or --number is required.");
}
function parseWorkflowRunEvent(raw) {
	if (typeof raw !== "object" || raw === null || !("workflow_run" in raw)) return {
		ok: false,
		reason: "A workflow_run event is required."
	};
	const workflowRun = raw.workflow_run;
	if (typeof workflowRun !== "object" || workflowRun === null) return {
		ok: false,
		reason: "A workflow_run event is required."
	};
	if (workflowRun.event !== "pull_request") return {
		ok: false,
		reason: "The preview workflow was not triggered by a pull request."
	};
	if (workflowRun.conclusion !== "success") return {
		ok: false,
		reason: "The preview workflow did not complete successfully."
	};
	const pullRequests = workflowRun.pull_requests;
	if (!Array.isArray(pullRequests) || pullRequests.length === 0) return {
		ok: false,
		reason: "The preview workflow is not associated with a pull request."
	};
	const number = pullRequests[0].number;
	if (typeof number !== "number" || !Number.isInteger(number)) return {
		ok: false,
		reason: "The preview workflow is not associated with a pull request."
	};
	return {
		ok: true,
		number
	};
}
async function readPullRequestFromWorkflowRunEvent() {
	const eventPath = process.env.GITHUB_EVENT_PATH;
	if (!eventPath) return {
		found: false,
		reason: "GITHUB_EVENT_PATH is required."
	};
	let raw;
	try {
		raw = JSON.parse(await readFile(eventPath, "utf8"));
	} catch {
		return {
			found: false,
			reason: "Failed to read workflow_run event."
		};
	}
	const parsed = parseWorkflowRunEvent(raw);
	if (!parsed.ok) return {
		found: false,
		reason: parsed.reason
	};
	return {
		found: true,
		number: parsed.number
	};
}
async function listPullRequestChangelogFiles(context, baseSha, headSha) {
	const dir = relative(context.cwd, context.changelogDir);
	const result = await x("git", [
		"diff",
		"--name-only",
		"--diff-filter=ACMRD",
		`${baseSha}...${headSha}`,
		"--",
		`${dir}/`
	], { nodeOptions: { cwd: context.cwd } });
	if (result.exitCode !== 0) throw execFailure("Failed to list pull request changelog files.", result);
	const files = /* @__PURE__ */ new Set();
	for (const line of result.stdout.split("\n")) {
		const trimmed = line.trim();
		if (trimmed.endsWith(".md")) files.add(basename(trimmed));
	}
	return files;
}
function createChangelogUrl(context, repo, branch, filename) {
	const filePath = join(relative(context.cwd, context.changelogDir), filename).replaceAll("\\", "/");
	return `https://github.com/${repo}/new/${branch.split("/").map(encodeURIComponent).join("/")}?${new URLSearchParams({ filename: filePath })}`;
}
function parsePositiveInt(value, option) {
	const number = Number(value);
	if (!Number.isInteger(number) || number <= 0) throw new Error(`${option} must be a positive integer.`);
	return number;
}
//#endregion
//#region src/plugins/github.ts
/** Create GitHub releases for successfully published packages after the whole plan succeeds. */
function github(options = {}) {
	const { release: releaseOptions = true } = options;
	let autoRelease;
	return [
		git(options),
		{
			name: "github",
			init() {
				const { repo, token } = this.github = {
					repo: options.repo ?? process.env.GITHUB_REPOSITORY,
					token: options.token ?? process.env.GITHUB_TOKEN ?? process.env.GH_TOKEN
				};
				if (repo && token && releaseOptions !== false) {
					const { eager = false, create, createGrouped } = releaseOptions === true ? {} : releaseOptions;
					const resolveEntryMeta = cached((entry) => entry.id, async (entry) => {
						const commit = await resolveFileCommit(this, entry.filename);
						if (!commit) return {
							commit,
							pullRequests: []
						};
						return {
							commit,
							pullRequests: await listPullRequestsForCommit(repo, commit, token).catch(() => [])
						};
					});
					autoRelease = createAutoRelease({
						eager,
						override: create,
						overrideGroup: createGrouped,
						async formatChangelog(entry) {
							const meta = await resolveEntryMeta(entry);
							const lines = [];
							const commitSuffix = meta.commit ? ` ([${meta.commit.slice(0, 7)}](https://github.com/${repo}/commit/${meta.commit}))` : "";
							for (const section of entry.sections) {
								lines.push(`### ${section.title}${commitSuffix}`, "");
								if (section.content) lines.push(section.content, "");
							}
							if (meta.pullRequests.length > 0) {
								lines.push("<details>", "<summary>Pull request & contributors</summary>", "");
								for (const pr of meta.pullRequests) {
									let line = `- [#${pr.number} ${pr.title}](https://github.com/${repo}/pull/${pr.number})`;
									if (pr.user) line += ` by @${pr.user.login}`;
									lines.push(line);
								}
								lines.push("", "</details>");
							}
							return lines.join("\n").trim();
						},
						create({ input, packages, tag }) {
							return createRelease({
								...input,
								repo,
								tag,
								token,
								prerelease: input.prerelease ?? packages.some((pkg) => pkg.version && semver.prerelease(pkg.version))
							});
						},
						releaseExistsByTag(tag) {
							return releaseExistsByTag(repo, tag, token);
						}
					});
				}
			},
			async resolvePlanStatus({ plan }) {
				if (await autoRelease?.hasPending.call(this, plan)) return "pending";
			},
			async afterPublishAll({ plan }) {
				await autoRelease?.create.call(this, plan);
			},
			async initCli(cli) {
				registerPrCli(cli);
				if (!isCI()) return;
				const { repo, token } = this.github ?? {};
				if (!token || !repo) return;
				const result = await x("git", [
					"remote",
					"set-url",
					"origin",
					`https://x-access-token:${token}@github.com/${repo}.git`
				], { nodeOptions: { cwd: this.cwd } });
				if (result.exitCode !== 0) throw execFailure("Failed to configure git remote for GitHub Actions.", result);
			}
		},
		versionRequestPlugin({
			name: "github",
			options: options.versionPr,
			canCreate(context) {
				const { repo, token } = context.github ?? {};
				return Boolean(repo && token);
			},
			async upsert(context, request, update) {
				const { repo, token } = context.github;
				const openPr = await findOpenPullRequest(repo, request.head, token);
				if (openPr === void 0) await createPullRequest(repo, {
					title: request.title,
					body: request.body,
					head: request.head,
					base: request.base,
					token
				});
				else if (update) await updatePullRequest(repo, openPr, {
					title: request.title,
					body: request.body,
					token
				});
			}
		})
	];
}
//#endregion
export { github };
