// tegami/dist/plugins/gitlab.js
import { t as changelogFilename } from "../generate-Cip5o2xy.js";
import { c as joinPath, n as execFailure, o as cached, r as fetchFailure, s as isCI } from "../error-CAsrGb6e.js";
import { a as createDraft, l as readChangelogEntries } from "../publish-CQifflae.js";
import { git } from "./git.js";
import { i as versionRequestPlugin, n as formatPreview, r as resolveFileCommit, t as createAutoRelease } from "../version-request-CLEmlVnL.js";
import { readFile, writeFile } from "node:fs/promises";
import { basename, join, relative, resolve } from "node:path";
import { x } from "tinyexec";
import { note, outro } from "@clack/prompts";
//#region src/plugins/gitlab/api.ts
function parseGitLabRepo(repo) {
	const projectPath = repo.replace(/^\/+|\/+$/g, "");
	if (!projectPath || !projectPath.includes("/")) throw new Error(`Invalid GitLab repository: ${repo}`);
	return {
		projectPath,
		encodedProjectPath: encodeURIComponent(projectPath)
	};
}
async function gitlabRequest(options, path, init = {}) {
	const headers = new Headers(init.headers);
	headers.set("Accept", "application/json");
	if (options.token) headers.set(options.token.type === "job-token" ? "JOB-TOKEN" : "PRIVATE-TOKEN", options.token.value);
	return fetch(joinPath(options.apiUrl ?? "https://gitlab.com/api/v4", path), {
		...init,
		headers
	});
}
async function releaseExistsByTag(repo, tag, options = {}) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/releases/${encodeURIComponent(tag)}`, { method: "HEAD" });
	if (response.status === 404) return false;
	if (!response.ok) throw await fetchFailure(`Failed to get GitLab release for ${tag}`, response);
	return true;
}
async function createRelease(repo, options) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/releases`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			tag_name: options.tag,
			name: options.title,
			description: options.notes
		})
	});
	if (!response.ok) throw await fetchFailure(`Failed to create GitLab release for ${options.tag}`, response);
}
async function findOpenMergeRequest(repo, options) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const params = new URLSearchParams({
		source_branch: options.head,
		state: "opened"
	});
	if (options.base) params.set("target_branch", options.base);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests?${params}`);
	if (!response.ok) throw await fetchFailure("Failed to check for an existing version merge request", response);
	return (await response.json())[0]?.iid;
}
async function updateMergeRequest(repo, number, options) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests/${number}`, {
		method: "PUT",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			title: options.title,
			description: options.body,
			target_branch: options.base
		})
	});
	if (!response.ok) throw await fetchFailure("Failed to update the version merge request", response);
}
async function createMergeRequest(repo, options) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			title: options.title,
			description: options.body,
			source_branch: options.head,
			target_branch: options.base
		})
	});
	if (!response.ok) throw await fetchFailure("Failed to create the version merge request", response);
}
async function getMergeRequest(repo, number, options = {}) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests/${number}`);
	if (!response.ok) throw await fetchFailure(`Failed to resolve merge request !${number}`, response);
	const data = await response.json();
	let sourceProjectPath;
	if (data.source_project_id !== void 0) {
		const projectResponse = await gitlabRequest(options, `/projects/${data.source_project_id}`);
		if (projectResponse.ok) sourceProjectPath = (await projectResponse.json()).path_with_namespace;
	}
	return {
		sourceBranch: data.source_branch,
		sourceProjectPath,
		baseSha: data.diff_refs?.base_sha ?? "",
		headSha: data.diff_refs?.head_sha ?? data.sha ?? ""
	};
}
async function listMergeRequestsForCommit(repo, commitSha, options = {}) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/repository/commits/${commitSha}/merge_requests`);
	if (!response.ok) throw await fetchFailure(`Failed to list merge requests for commit ${commitSha.slice(0, 7)}`, response);
	return (await response.json()).map((mergeRequest) => ({
		number: mergeRequest.iid,
		title: mergeRequest.title,
		user: mergeRequest.author ? { login: mergeRequest.author.username } : void 0
	}));
}
async function findMergeRequestCommentByPrefix(repo, mergeRequestNumber, prefix, options = {}) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	let page = 1;
	while (true) {
		const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests/${mergeRequestNumber}/notes?per_page=100&page=${page}`);
		if (!response.ok) throw await fetchFailure("Failed to list merge request comments", response);
		const batch = await response.json();
		if (batch.length === 0) break;
		const comment = batch.find((comment) => comment.body.startsWith(prefix));
		if (comment) return comment.id;
		if (batch.length < 100) break;
		page += 1;
	}
}
async function updateMergeRequestComment(repo, mergeRequestNumber, commentId, body, options = {}) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests/${mergeRequestNumber}/notes/${commentId}`, {
		method: "PUT",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ body })
	});
	if (!response.ok) throw await fetchFailure("Failed to update merge request comment", response);
}
async function createMergeRequestComment(repo, mergeRequestNumber, body, options = {}) {
	const { encodedProjectPath } = parseGitLabRepo(repo);
	const response = await gitlabRequest(options, `/projects/${encodedProjectPath}/merge_requests/${mergeRequestNumber}/notes`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ body })
	});
	if (!response.ok) throw await fetchFailure("Failed to create merge request comment", response);
}
//#endregion
//#region src/plugins/gitlab/cli.ts
const COMMENT_MARKER = "<!-- tegami -->";
function registerMrCli(cli) {
	cli.command("mr preview", { description: "show a merge request release preview and changelog guidance" }).option("artifact", {
		type: "string",
		description: "write preview markdown to a file"
	}).option("number", {
		type: "string",
		description: "merge request iid"
	}).action(async ({ context, values }) => {
		const number = values.number ? parsePositiveInt(values.number, "--number") : void 0;
		const artifact = values.artifact;
		const body = await buildMrPreview(context, await createDraft(await readChangelogEntries(context), context), { number });
		if (artifact) {
			const artifactPath = resolve(context.cwd, artifact);
			await writeFile(artifactPath, body);
			if (!isCI()) {
				note(relative(context.cwd, artifactPath) || artifact, "Release preview");
				outro("Release preview ready.");
			}
			return;
		}
		if (isCI()) {
			process.stdout.write(`${body}\n`);
			return;
		}
		note(body, "Release preview");
		outro("Release preview ready.");
	});
	cli.command("mr comment", { description: "post the merge request release preview as a comment" }).positional("artifact").option("number", {
		type: "string",
		description: "merge request iid"
	}).action(async ({ context, values, positionals }) => {
		await postMrComment(context, await readFile(positionals.artifact, "utf8"), { number: values.number ? parsePositiveInt(values.number, "--number") : void 0 });
		outro("Merge request comment updated.");
	});
}
async function buildMrPreview(context, draft, options = {}) {
	const mergeRequest = await resolveMergeRequest(context, options);
	return formatPreview(context, draft, await listMergeRequestChangelogFiles(context, mergeRequest.baseSha, mergeRequest.headSha), {
		"create-a-changelog-href": createChangelogUrl(context, mergeRequest.headRepo, mergeRequest.headRef, changelogFilename()),
		pr: "MR"
	});
}
async function postMrComment(context, body, options = {}) {
	const { repo, apiUrl, token } = context.gitlab;
	if (!repo) {
		outro("GITLAB_REPOSITORY or CI_PROJECT_PATH is required.");
		return;
	}
	const number = options.number ?? (process.env.CI_MERGE_REQUEST_IID ? parsePositiveInt(process.env.CI_MERGE_REQUEST_IID, "CI_MERGE_REQUEST_IID") : void 0);
	if (!number) {
		outro("CI_MERGE_REQUEST_IID or --number is required.");
		return;
	}
	const api = {
		apiUrl,
		token
	};
	const markedBody = `${COMMENT_MARKER}\n${body}`;
	const existingId = await findMergeRequestCommentByPrefix(repo, number, COMMENT_MARKER, api);
	if (existingId) {
		await updateMergeRequestComment(repo, number, existingId, markedBody, api);
		return;
	}
	await createMergeRequestComment(repo, number, markedBody, api);
}
async function resolveMergeRequest(context, options) {
	const { repo, apiUrl, token } = context.gitlab;
	if (!repo) throw new Error("GITLAB_REPOSITORY or CI_PROJECT_PATH is required.");
	const api = {
		apiUrl,
		token
	};
	if (options.number !== void 0) {
		const data = await getMergeRequest(repo, options.number, api);
		if (!data.baseSha || !data.headSha) throw new Error(`Merge request !${options.number} does not include diff refs.`);
		return {
			headRepo: data.sourceProjectPath ?? repo,
			headRef: data.sourceBranch,
			baseSha: data.baseSha,
			headSha: data.headSha
		};
	}
	const headRef = process.env.CI_MERGE_REQUEST_SOURCE_BRANCH_NAME;
	const baseSha = process.env.CI_MERGE_REQUEST_DIFF_BASE_SHA;
	const headSha = process.env.CI_COMMIT_SHA;
	if (headRef && baseSha && headSha) return {
		headRepo: process.env.CI_MERGE_REQUEST_SOURCE_PROJECT_PATH ?? repo,
		headRef,
		baseSha,
		headSha
	};
	throw new Error("A merge request event or --number is required.");
}
async function listMergeRequestChangelogFiles(context, baseSha, headSha) {
	const dir = relative(context.cwd, context.changelogDir);
	const result = await x("git", [
		"diff",
		"--name-only",
		"--diff-filter=ACMRD",
		`${baseSha}...${headSha}`,
		"--",
		`${dir}/`
	], { nodeOptions: { cwd: context.cwd } });
	if (result.exitCode !== 0) throw execFailure("Failed to list merge request changelog files.", result);
	const files = /* @__PURE__ */ new Set();
	for (const line of result.stdout.split("\n")) {
		const trimmed = line.trim();
		if (trimmed.endsWith(".md")) files.add(basename(trimmed));
	}
	return files;
}
function createChangelogUrl(context, repo, branch, filename) {
	const filePath = join(relative(context.cwd, context.changelogDir), filename).replaceAll("\\", "/");
	const branchPath = branch.split("/").map(encodeURIComponent).join("/");
	const params = new URLSearchParams({ file_name: filePath });
	return joinPath(context.gitlab.webUrl, repo, "-/new", branchPath) + `?${params}`;
}
function parsePositiveInt(value, option) {
	const number = Number(value);
	if (!Number.isInteger(number) || number <= 0) throw new Error(`${option} must be a positive integer.`);
	return number;
}
//#endregion
//#region src/plugins/gitlab.ts
/** Create GitLab releases for successfully published packages after the whole plan succeeds. */
function gitlab(options = {}) {
	const { release: releaseOptions = true } = options;
	let autoRelease;
	return [
		git(options),
		{
			name: "gitlab",
			init() {
				const { repo, token, webUrl } = this.gitlab = {
					repo: options.repo ?? process.env.GITLAB_REPOSITORY ?? process.env.CI_PROJECT_PATH,
					token: resolveGitLabToken(options.token),
					apiUrl: options.apiUrl ?? process.env.GITLAB_API_URL ?? process.env.CI_API_V4_URL ?? "https://gitlab.com/api/v4",
					webUrl: options.webUrl ?? process.env.GITLAB_SERVER_URL ?? process.env.CI_SERVER_URL ?? "https://gitlab.com"
				};
				if (repo && token && releaseOptions !== false) {
					const { eager = false, create, createGrouped } = releaseOptions === true ? {} : releaseOptions;
					const api = gitLabApiOptions(this.gitlab);
					const resolveEntryMeta = cached((entry) => entry.id, async (entry) => {
						const commit = await resolveFileCommit(this, entry.filename);
						if (!commit || !repo) return {
							commit,
							mergeRequests: []
						};
						return {
							commit,
							mergeRequests: await listMergeRequestsForCommit(repo, commit, api).catch(() => [])
						};
					});
					autoRelease = createAutoRelease({
						eager,
						override: create,
						overrideGroup: createGrouped,
						async formatChangelog(entry) {
							const meta = await resolveEntryMeta(entry);
							const commitSuffix = meta.commit ? ` ([${meta.commit.slice(0, 7)}](${joinPath(webUrl, repo, "-/commit", meta.commit)}))` : "";
							const lines = [];
							for (const section of entry.sections) {
								lines.push(`### ${section.title}${commitSuffix}`, "");
								if (section.content) lines.push(section.content, "");
							}
							if (meta.mergeRequests.length > 0) {
								lines.push("<details>", "<summary>Merge request & contributors</summary>", "");
								for (const mr of meta.mergeRequests) {
									let line = `- [!${mr.number} ${mr.title}](${joinPath(webUrl, repo, "-/merge_requests", String(mr.number))})`;
									if (mr.user) line += ` by @${mr.user.login}`;
									lines.push(line);
								}
								lines.push("", "</details>");
							}
							return lines.join("\n").trim();
						},
						create({ input, tag }) {
							return createRelease(repo, {
								tag,
								title: input.title,
								notes: input.notes,
								...api
							});
						},
						releaseExistsByTag(tag) {
							return releaseExistsByTag(repo, tag, api);
						}
					});
				}
			},
			async resolvePlanStatus({ plan }) {
				if (await autoRelease?.hasPending.call(this, plan)) return "pending";
			},
			async afterPublishAll({ plan }) {
				await autoRelease?.create.call(this, plan);
			},
			async initCli(cli) {
				registerMrCli(cli);
				if (!isCI()) return;
				const { repo, token, webUrl } = this.gitlab;
				if (!token || !repo) return;
				const result = await x("git", [
					"remote",
					"set-url",
					"origin",
					gitlabRemoteUrl(repo, token, webUrl)
				], { nodeOptions: { cwd: this.cwd } });
				if (result.exitCode !== 0) throw execFailure("Failed to configure git remote for GitLab CI.", result);
			}
		},
		versionRequestPlugin({
			name: "gitlab",
			options: options.versionMr,
			canCreate(context) {
				const { repo, token } = context.gitlab ?? {};
				return Boolean(repo && token);
			},
			async upsert(context, request, update) {
				const repo = context.gitlab.repo;
				const api = gitLabApiOptions(context.gitlab);
				const openMr = await findOpenMergeRequest(repo, {
					head: request.head,
					base: request.base,
					...api
				});
				if (openMr === void 0) await createMergeRequest(repo, {
					title: request.title,
					body: request.body,
					head: request.head,
					base: request.base,
					...api
				});
				else if (update) await updateMergeRequest(repo, openMr, {
					title: request.title,
					body: request.body,
					base: request.base,
					...api
				});
			}
		})
	];
}
function gitLabApiOptions(gitlab) {
	const options = {};
	if (gitlab?.apiUrl) options.apiUrl = gitlab.apiUrl;
	if (gitlab?.token) options.token = gitlab.token;
	return options;
}
function resolveGitLabToken(optionToken) {
	if (optionToken) return {
		value: optionToken,
		type: "private-token"
	};
	if (process.env.GITLAB_TOKEN) return {
		value: process.env.GITLAB_TOKEN,
		type: "private-token"
	};
	if (process.env.GL_TOKEN) return {
		value: process.env.GL_TOKEN,
		type: "private-token"
	};
	if (process.env.CI_JOB_TOKEN) return {
		value: process.env.CI_JOB_TOKEN,
		type: "job-token"
	};
}
function gitlabRemoteUrl(repo, token, webUrl) {
	const username = token.type === "job-token" ? "gitlab-ci-token" : "oauth2";
	return joinPath(webUrl.replace(/^https?:\/\//, `https://${username}:${token.value}@`), `${repo}.git`);
}
//#endregion
export { gitlab };
