// tegami/dist/plugins/go.js
import { a as isNodeError, n as execFailure, r as fetchFailure } from "../error-CAsrGb6e.js";
import { n as _validateReport, t as _createStandardSchema } from "../_createStandardSchema-DFfVlh8x.js";
import { n as WorkspacePackage } from "../graph-DaJ28Y4f.js";
import { relative, resolve } from "node:path";
import { x } from "tinyexec";
import * as semver$1 from "semver";
//#region src/plugins/go.ts
const validateGoWorkJson = (() => {
	const _io0 = (input) => void 0 === input.Use || Array.isArray(input.Use) && input.Use.every((elem) => "object" === typeof elem && null !== elem && _io1(elem));
	const _io1 = (input) => "string" === typeof input.DiskPath && (void 0 === input.ModulePath || "string" === typeof input.ModulePath);
	const _vo0 = (input, _path, _exceptionable = true) => [void 0 === input.Use || (Array.isArray(input.Use) || _report(_exceptionable, {
		path: _path + ".Use",
		expected: "(Array<__type> | undefined)",
		value: input.Use
	})) && input.Use.map((elem, _index2) => ("object" === typeof elem && null !== elem || _report(_exceptionable, {
		path: _path + ".Use[" + _index2 + "]",
		expected: "__type",
		value: elem
	})) && _vo1(elem, _path + ".Use[" + _index2 + "]", _exceptionable) || _report(_exceptionable, {
		path: _path + ".Use[" + _index2 + "]",
		expected: "__type",
		value: elem
	})).every((flag) => flag) || _report(_exceptionable, {
		path: _path + ".Use",
		expected: "(Array<__type> | undefined)",
		value: input.Use
	})].every((flag) => flag);
	const _vo1 = (input, _path, _exceptionable = true) => ["string" === typeof input.DiskPath || _report(_exceptionable, {
		path: _path + ".DiskPath",
		expected: "string",
		value: input.DiskPath
	}), void 0 === input.ModulePath || "string" === typeof input.ModulePath || _report(_exceptionable, {
		path: _path + ".ModulePath",
		expected: "(string | undefined)",
		value: input.ModulePath
	})].every((flag) => flag);
	const __is = (input) => "object" === typeof input && null !== input && false === Array.isArray(input) && _io0(input);
	let errors;
	let _report;
	return _createStandardSchema((input) => {
		if (false === __is(input)) {
			errors = [];
			_report = _validateReport(errors);
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input && false === Array.isArray(input) || _report(true, {
				path: _path + "",
				expected: "GoWorkJson",
				value: input
			})) && _vo0(input, _path + "", true) || _report(true, {
				path: _path + "",
				expected: "GoWorkJson",
				value: input
			}))(input, "$input", true);
			const success = 0 === errors.length;
			return success ? {
				success,
				data: input
			} : {
				success,
				errors,
				data: input
			};
		}
		return {
			success: true,
			data: input
		};
	});
})();
const validateGoModJson = (() => {
	const _io0 = (input) => "object" === typeof input.Module && null !== input.Module && _io1(input.Module) && (void 0 === input.Require || Array.isArray(input.Require) && input.Require.every((elem) => "object" === typeof elem && null !== elem && _io2(elem))) && (void 0 === input.Replace || Array.isArray(input.Replace) && input.Replace.every((elem) => "object" === typeof elem && null !== elem && _io3(elem)));
	const _io1 = (input) => "string" === typeof input.Path;
	const _io2 = (input) => "string" === typeof input.Path && (void 0 === input.Version || "string" === typeof input.Version);
	const _io3 = (input) => "object" === typeof input.Old && null !== input.Old && _io4(input.Old) && "object" === typeof input.New && null !== input.New && _io5(input.New);
	const _io4 = (input) => "string" === typeof input.Path && (void 0 === input.Version || "string" === typeof input.Version);
	const _io5 = (input) => "string" === typeof input.Path && (void 0 === input.Version || "string" === typeof input.Version);
	const _vo0 = (input, _path, _exceptionable = true) => [
		("object" === typeof input.Module && null !== input.Module || _report(_exceptionable, {
			path: _path + ".Module",
			expected: "__type",
			value: input.Module
		})) && _vo1(input.Module, _path + ".Module", _exceptionable) || _report(_exceptionable, {
			path: _path + ".Module",
			expected: "__type",
			value: input.Module
		}),
		void 0 === input.Require || (Array.isArray(input.Require) || _report(_exceptionable, {
			path: _path + ".Require",
			expected: "(Array<__type> | undefined)",
			value: input.Require
		})) && input.Require.map((elem, _index3) => ("object" === typeof elem && null !== elem || _report(_exceptionable, {
			path: _path + ".Require[" + _index3 + "]",
			expected: "__type.o1",
			value: elem
		})) && _vo2(elem, _path + ".Require[" + _index3 + "]", _exceptionable) || _report(_exceptionable, {
			path: _path + ".Require[" + _index3 + "]",
			expected: "__type.o1",
			value: elem
		})).every((flag) => flag) || _report(_exceptionable, {
			path: _path + ".Require",
			expected: "(Array<__type> | undefined)",
			value: input.Require
		}),
		void 0 === input.Replace || (Array.isArray(input.Replace) || _report(_exceptionable, {
			path: _path + ".Replace",
			expected: "(Array<__type>.o1 | undefined)",
			value: input.Replace
		})) && input.Replace.map((elem, _index4) => ("object" === typeof elem && null !== elem || _report(_exceptionable, {
			path: _path + ".Replace[" + _index4 + "]",
			expected: "__type.o2",
			value: elem
		})) && _vo3(elem, _path + ".Replace[" + _index4 + "]", _exceptionable) || _report(_exceptionable, {
			path: _path + ".Replace[" + _index4 + "]",
			expected: "__type.o2",
			value: elem
		})).every((flag) => flag) || _report(_exceptionable, {
			path: _path + ".Replace",
			expected: "(Array<__type>.o1 | undefined)",
			value: input.Replace
		})
	].every((flag) => flag);
	const _vo1 = (input, _path, _exceptionable = true) => ["string" === typeof input.Path || _report(_exceptionable, {
		path: _path + ".Path",
		expected: "string",
		value: input.Path
	})].every((flag) => flag);
	const _vo2 = (input, _path, _exceptionable = true) => ["string" === typeof input.Path || _report(_exceptionable, {
		path: _path + ".Path",
		expected: "string",
		value: input.Path
	}), void 0 === input.Version || "string" === typeof input.Version || _report(_exceptionable, {
		path: _path + ".Version",
		expected: "(string | undefined)",
		value: input.Version
	})].every((flag) => flag);
	const _vo3 = (input, _path, _exceptionable = true) => [("object" === typeof input.Old && null !== input.Old || _report(_exceptionable, {
		path: _path + ".Old",
		expected: "__type.o3",
		value: input.Old
	})) && _vo4(input.Old, _path + ".Old", _exceptionable) || _report(_exceptionable, {
		path: _path + ".Old",
		expected: "__type.o3",
		value: input.Old
	}), ("object" === typeof input.New && null !== input.New || _report(_exceptionable, {
		path: _path + ".New",
		expected: "__type.o4",
		value: input.New
	})) && _vo5(input.New, _path + ".New", _exceptionable) || _report(_exceptionable, {
		path: _path + ".New",
		expected: "__type.o4",
		value: input.New
	})].every((flag) => flag);
	const _vo4 = (input, _path, _exceptionable = true) => ["string" === typeof input.Path || _report(_exceptionable, {
		path: _path + ".Path",
		expected: "string",
		value: input.Path
	}), void 0 === input.Version || "string" === typeof input.Version || _report(_exceptionable, {
		path: _path + ".Version",
		expected: "(string | undefined)",
		value: input.Version
	})].every((flag) => flag);
	const _vo5 = (input, _path, _exceptionable = true) => ["string" === typeof input.Path || _report(_exceptionable, {
		path: _path + ".Path",
		expected: "string",
		value: input.Path
	}), void 0 === input.Version || "string" === typeof input.Version || _report(_exceptionable, {
		path: _path + ".Version",
		expected: "(string | undefined)",
		value: input.Version
	})].every((flag) => flag);
	const __is = (input) => "object" === typeof input && null !== input && _io0(input);
	let errors;
	let _report;
	return _createStandardSchema((input) => {
		if (false === __is(input)) {
			errors = [];
			_report = _validateReport(errors);
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input || _report(true, {
				path: _path + "",
				expected: "GoModJson",
				value: input
			})) && _vo0(input, _path + "", true) || _report(true, {
				path: _path + "",
				expected: "GoModJson",
				value: input
			}))(input, "$input", true);
			const success = 0 === errors.length;
			return success ? {
				success,
				data: input
			} : {
				success,
				errors,
				data: input
			};
		}
		return {
			success: true,
			data: input
		};
	});
})();
var GoPackage = class extends WorkspacePackage {
	path;
	mod;
	manager = "go";
	versionValue;
	pendingRequires = /* @__PURE__ */ new Map();
	constructor(path, mod, version) {
		super();
		this.path = path;
		this.mod = mod;
		this.versionValue = version;
	}
	get name() {
		return this.mod.module;
	}
	get version() {
		return this.versionValue;
	}
	setVersion(version) {
		this.versionValue = version;
	}
	setRequire(module, version) {
		const formatted = formatGoVersion(version);
		this.mod.requires.set(module, formatted);
		this.pendingRequires.set(module, formatted);
	}
	async write() {
		if (this.pendingRequires.size === 0) return;
		const args = ["mod", "edit"];
		for (const [module, version] of this.pendingRequires) args.push(`-require=${module}@${version}`);
		const result = await x("go", args, { nodeOptions: { cwd: this.path } });
		if (result.exitCode !== 0) throw execFailure(`Failed to update go.mod requires in ${this.path}.`, result);
		this.pendingRequires.clear();
	}
};
const validateGoPackageLock = (() => {
	const _io0 = (input) => "string" === typeof input.id && "string" === typeof input.version;
	const _vo0 = (input, _path, _exceptionable = true) => ["string" === typeof input.id || _report(_exceptionable, {
		path: _path + ".id",
		expected: "string",
		value: input.id
	}), "string" === typeof input.version || _report(_exceptionable, {
		path: _path + ".version",
		expected: "string",
		value: input.version
	})].every((flag) => flag);
	const __is = (input) => "object" === typeof input && null !== input && _io0(input);
	let errors;
	let _report;
	return _createStandardSchema((input) => {
		if (false === __is(input)) {
			errors = [];
			_report = _validateReport(errors);
			((input, _path, _exceptionable = true) => ("object" === typeof input && null !== input || _report(true, {
				path: _path + "",
				expected: "GoPackageLock",
				value: input
			})) && _vo0(input, _path + "", true) || _report(true, {
				path: _path + "",
				expected: "GoPackageLock",
				value: input
			}))(input, "$input", true);
			const success = 0 === errors.length;
			return success ? {
				success,
				data: input
			} : {
				success,
				errors,
				data: input
			};
		}
		return {
			success: true,
			data: input
		};
	});
})();
/**
* Plugin for Golang, the release flow of Golang is pretty special, there's some exceptions for it:
*
* - Version is stored in lock file, normally, it should prefer to store versions in a file like `package.json` & `Cargo.toml`, but for Golang, there is no such file.
* - Publishing is handed to Git plugin, because Golang uses Git tag for publishing.
*/
function go({ updateLockFile = true, bumpDep: getBumpDepType } = {}) {
	let active = false;
	return {
		name: "go",
		async resolve() {
			await discoverGoPackages(this.cwd, (pkg) => this.graph.add(pkg));
			active = this.graph.getPackages().some((pkg) => pkg instanceof GoPackage);
			if (active && !this.plugins.some((plugin) => plugin.name === "git")) throw new Error("The go plugin requires the git plugin. Add git() from \"tegami/plugins/git\" to your plugins array.");
		},
		initDraft(plan) {
			if (!active) return;
			plan.addPolicy(depsPolicy(this, getBumpDepType));
		},
		async applyDraft(draft) {
			if (!active) return;
			const { graph } = this;
			const writes = [];
			for (const pkg of graph.getPackages()) {
				if (!(pkg instanceof GoPackage)) continue;
				const bumped = draft.getPackageDraft(pkg.id)?.bumpVersion(pkg);
				if (bumped) pkg.setVersion(bumped);
			}
			for (const pkg of graph.getPackages()) {
				if (!(pkg instanceof GoPackage)) continue;
				for (const [moduleName, requireVersion] of pkg.mod.requires) {
					const linked = graph.get(`go:${moduleName}`);
					if (!linked || !(linked instanceof GoPackage)) continue;
					if (semver$1.satisfies(linked.version, stripGoVersion(requireVersion))) continue;
					pkg.setRequire(moduleName, linked.version);
				}
				writes.push(pkg.write());
			}
			await Promise.all(writes);
		},
		initPublishLock({ lock, draft }) {
			if (!active) return;
			for (const id of draft.getPackageDrafts().keys()) {
				const pkg = this.graph.get(id);
				if (!(pkg instanceof GoPackage)) continue;
				lock.write("go:packages", {
					id,
					version: pkg.version
				});
			}
		},
		initPublishPlan({ lock, plan }) {
			if (!active) return;
			const lockVersions = /* @__PURE__ */ new Map();
			let data;
			while (data = lock.read("go:packages")) {
				const { success, data: parsed } = validateGoPackageLock(data);
				if (!success) continue;
				lockVersions.set(parsed.id, parsed.version);
			}
			for (const [id, packagePlan] of plan.packages) {
				const pkg = this.graph.get(id);
				if (!(pkg instanceof GoPackage)) continue;
				const version = packagePlan.updated && lockVersions.get(id);
				if (version) pkg.setVersion(version);
				packagePlan.git ??= {};
				packagePlan.git.tag = formatGoTag(this.cwd, pkg.path, pkg.version);
			}
		},
		async publishPreflight({ pkg }) {
			if (!(pkg instanceof GoPackage)) return;
			const shouldPublish = pkg.options.go?.publish ?? pkg.group?.options?.go?.publish ?? true;
			const wait = [];
			for (const [moduleName] of pkg.mod.requires) {
				const linked = this.graph.get(`go:${moduleName}`);
				if (linked instanceof GoPackage) wait.push(linked.id);
			}
			for (const replace of pkg.mod.replaces.values()) {
				if (!replace.path) continue;
				const linked = findLocalModule(this.graph, pkg.path, replace.path);
				if (!linked) continue;
				wait.push(linked.id);
			}
			return {
				shouldPublish,
				wait
			};
		},
		resolvePlanStatus({ plan }) {
			if (!active) return;
			return Array.from(plan.packages, async ([id, { preflight }]) => {
				if (!preflight.shouldPublish) return;
				const pkg = this.graph.get(id);
				if (!(pkg instanceof GoPackage)) return;
				if (!await isModulePublished(pkg.name, pkg.version)) return "pending";
			});
		},
		async publish({ pkg }) {
			if (!(pkg instanceof GoPackage)) return;
			return { type: await isModulePublished(pkg.name, pkg.version) ? "skipped" : "published" };
		},
		async applyCliDraft() {
			if (!active || !updateLockFile) return;
			if (await listGoWorkUsePaths(this.cwd)) {
				const result = await x("go", ["work", "sync"], { nodeOptions: { cwd: this.cwd } });
				if (result.exitCode !== 0) throw execFailure("Failed to run `go work sync`.", result);
				return;
			}
			await Promise.all(this.graph.getPackages().map(async (pkg) => {
				if (!(pkg instanceof GoPackage)) return;
				const result = await x("go", ["mod", "tidy"], { nodeOptions: { cwd: pkg.path } });
				if (result.exitCode !== 0) throw execFailure(`Failed to run \`go mod tidy\` in ${pkg.path}.`, result);
			}));
		}
	};
}
function depsPolicy({ graph }, getBumpDepType = () => "patch") {
	const dependentMap = /* @__PURE__ */ new Map();
	for (const pkg of graph.getPackages()) {
		if (!(pkg instanceof GoPackage)) continue;
		for (const [name, version] of pkg.mod.requires) {
			const id = `go:${name}`;
			const refs = dependentMap.get(id);
			if (refs) refs.push({
				dependent: pkg,
				name,
				version
			});
			else dependentMap.set(id, [{
				dependent: pkg,
				name,
				version
			}]);
		}
	}
	return {
		id: "go:deps",
		onUpdate({ pkg, packageDraft: plan }) {
			if (!(pkg instanceof GoPackage)) return;
			const deps = dependentMap.get(pkg.id);
			if (!deps) return;
			const bumped = plan.bumpVersion(pkg);
			if (!bumped) return;
			for (const dep of deps) {
				if (pkg.group?.options.syncBump && dep.dependent.group === pkg.group) continue;
				if (semver$1.satisfies(bumped, stripGoVersion(dep.version))) continue;
				const bumpType = getBumpDepType?.(dep);
				if (bumpType === false) continue;
				this.bumpPackage(dep.dependent, {
					type: bumpType,
					reason: `update require "${dep.name}"`
				});
			}
		}
	};
}
async function discoverGoPackages(cwd, add) {
	const modulePaths = await listModulePaths(cwd);
	await Promise.all(modulePaths.map(async (modulePath) => {
		const mod = await readGoMod(modulePath).catch((error) => {
			if (isNodeError(error) && error.code === "ENOENT") return void 0;
			throw error;
		});
		if (!mod) return;
		add(new GoPackage(modulePath, mod, await readLatestVersion(cwd, modulePath)));
	}));
}
async function listModulePaths(cwd) {
	const workPaths = await listGoWorkUsePaths(cwd);
	if (workPaths) return workPaths;
	if (await readGoMod(cwd)) return [cwd];
	return [];
}
async function listGoWorkUsePaths(cwd) {
	const result = await x("go", [
		"work",
		"edit",
		"-json"
	], { nodeOptions: { cwd } });
	if (result.exitCode !== 0) return void 0;
	const validated = validateGoWorkJson(JSON.parse(result.stdout));
	const data = validated.success ? validated.data : void 0;
	if (!data?.Use?.length) return [cwd];
	return data.Use.map((use) => resolve(cwd, use.DiskPath));
}
async function readGoMod(path) {
	const result = await x("go", [
		"mod",
		"edit",
		"-json"
	], { nodeOptions: { cwd: path } });
	if (result.exitCode !== 0) return;
	const validated = validateGoModJson(JSON.parse(result.stdout));
	const data = validated.success ? validated.data : void 0;
	if (!data) return;
	const requires = /* @__PURE__ */ new Map();
	for (const req of data.Require ?? []) if (req.Version) requires.set(req.Path, req.Version);
	const replaces = /* @__PURE__ */ new Map();
	for (const rep of data.Replace ?? []) replaces.set(rep.Old.Path, {
		module: rep.New.Path,
		path: rep.New.Version ? void 0 : rep.New.Path,
		version: rep.New.Version
	});
	return {
		module: data.Module.Path,
		requires,
		replaces
	};
}
function findLocalModule(graph, modulePath, replacePath) {
	const absolute = resolve(modulePath, replacePath);
	return graph.getPackages().find((pkg) => pkg instanceof GoPackage && pkg.path === absolute);
}
async function readLatestVersion(cwd, modulePath) {
	const result = await x("git", [
		"tag",
		"--list",
		formatGoTag(cwd, modulePath, "v*"),
		"--sort=-v:refname"
	], { nodeOptions: { cwd } });
	if (result.exitCode !== 0) return "0.0.0";
	for (const tag of result.stdout.split("\n")) {
		const version = parseTagVersion(tag.trim());
		if (version) return version;
	}
	return "0.0.0";
}
function parseTagVersion(tag) {
	const version = stripGoVersion(tag.slice(tag.lastIndexOf("/") + 1));
	if (semver$1.valid(version)) return version;
}
function formatGoTag(cwd, modulePath, version) {
	const relativeDir = relative(cwd, modulePath).replaceAll("\\", "/");
	if (relativeDir === "") return formatGoVersion(version);
	return `${relativeDir}/${formatGoVersion(version)}`;
}
function formatGoVersion(version) {
	return version.startsWith("v") ? version : `v${version}`;
}
function stripGoVersion(version) {
	return version.startsWith("v") ? version.slice(1) : version;
}
async function isModulePublished(module, version) {
	const formatted = formatGoVersion(version);
	const response = await fetch(`https://proxy.golang.org/${encodeURIComponent(module)}/@v/${formatted}.info`);
	if (response.status === 200) return true;
	if (response.status === 404) return false;
	throw await fetchFailure(`Unable to validate ${module}@${formatted} against proxy.golang.org`, response);
}
//#endregion
export { GoPackage, go };
